<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>GO Annotation Analysis &mdash; oaklib  documentation</title>
      <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/nbsphinx-code-cells.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
        <script src="../../../_static/copybutton.js?v=046fa432"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            oaklib
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../intro/index.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../guide/index.html">The OAK Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../packages/index.html">OAK Library Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../cli.html">Command Line</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../datamodels/index.html">Datamodels</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../howtos/index.html">How-To Guides</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../index.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../faq/index.html">FAQ</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">oaklib</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">GO Annotation Analysis</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/examples/Ontologies/GO/Annotation-Analysis.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="GO-Annotation-Analysis">
<h1>GO Annotation Analysis<a class="headerlink" href="#GO-Annotation-Analysis" title="Link to this heading"></a></h1>
<p>Analyzes the associations between genes and GO terms in the Gene Ontology (GO) database.</p>
<p>In particular, categorizes associations based on:</p>
<ul class="simple">
<li><p>whether they have been “retracted” (i.e., removed from the database)</p></li>
<li><p>whether they have been reviewed and accepted or rejected via IBA</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>import pandas as pd

from tests.test_implementations.test_robot_template import adapter

# a recent release plus an older one that may have retracted entries

RELEASES = [
    &quot;2024-11-03&quot;,
    &quot;2024-06-10&quot;,
    &quot;2020-01-01&quot;,
]

LATEST = RELEASES[0]
PREVIOUS = RELEASES[1:]
assert all(r &lt; LATEST for r in PREVIOUS)
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>NEW_CUTOFF = &quot;2024-06-01&quot;
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span># taxa to analyze

TAXA = [
    (&quot;human&quot;, &quot;goa_human&quot;, 9606),
    (&quot;Arabidopsis thaliana&quot;, &quot;tair&quot;, 3702),
    (&quot;yeast&quot;, &quot;sgd&quot;, 559292),
]
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>GAF_URL_TEMPLATE = &quot;https://release.geneontology.org/{date}/annotations/{name}.gaf.gz&quot;
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>from oaklib.datamodels.vocabulary import IS_A, PART_OF
</pre></div>
</div>
</div>
<section id="Create-an-OAK-adapter-for-the-GO-ontology">
<h2>Create an OAK adapter for the GO ontology<a class="headerlink" href="#Create-an-OAK-adapter-for-the-GO-ontology" title="Link to this heading"></a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>from oaklib import get_adapter

go = get_adapter(&quot;sqlite:obo:go&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"version_major": 2, "version_minor": 0, "model_id": "aaf16a6ab071468590f8ec5a8690270e"}</script></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>obsoletes = set(go.obsoletes())
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>binding_terms = set(go.descendants(&quot;GO:0005488&quot;, predicates=[IS_A]))
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>antislim_terms = set(go.subset_members(&quot;gocheck_do_not_annotate&quot;)).union(go.subset_members(&quot;gocheck_obsoletion_candidate&quot;))
non_informative = binding_terms.union(antislim_terms)
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[96]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>mf_terms = set(go.descendants(&quot;GO:0003674&quot;, predicates=[IS_A]))
bp_terms = set(go.descendants(&quot;GO:0008150&quot;, predicates=[IS_A]))
cc_terms = set(go.descendants(&quot;GO:0005575&quot;, predicates=[IS_A]))
</pre></div>
</div>
</div>
</section>
<section id="Load-annotations-from-the-archive">
<h2>Load annotations from the archive<a class="headerlink" href="#Load-annotations-from-the-archive" title="Link to this heading"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>from oaklib.parsers import GafAssociationParser
gaf_parser = GafAssociationParser()
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>import requests_cache

session = requests_cache.CachedSession(
    cache_name=&#39;gaf_cache&#39;,
    backend=&#39;sqlite&#39;,  # or &#39;memory&#39; for in-memory cache
    expire_after=24*60*60,  # Cache expiration in seconds
    allowable_codes=[200],  # Only cache successful responses
)
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>from oaklib.datamodels.association import ParserConfiguration, NegatedAssociation
import io
import gzip

def get_gaf(release, name):
    config = ParserConfiguration(preserve_negated_associations=True)
    url = GAF_URL_TEMPLATE.format(date=release, name=name)
    # open the URL as a file object using requests
    with session.get(url, stream=True) as response:
        # Decompress the gzipped content and create a text stream
        decompressed = gzip.decompress(response.content)
        text_stream = io.TextIOWrapper(io.BytesIO(decompressed))
        print(f&quot;Reading {url} using {config}&quot;)
        return list(gaf_parser.parse(text_stream, configuration=config))
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>
<section id="Load-all-annotations-into-a-cache">
<h2>Load all annotations into a cache<a class="headerlink" href="#Load-all-annotations-into-a-cache" title="Link to this heading"></a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>from collections import defaultdict

db = defaultdict(dict)
for r in RELEASES:
    for name, grp, tax_id in TAXA:
        print(f&quot;Loading {r} {name}&quot;)
        assocs = get_gaf(r, grp)
        print(f&quot;Loaded {len(assocs)} associations&quot;)
        neg_assocs = [x for x in assocs if x.negated]
        print(f&quot;  {len(neg_assocs)} negated associations&quot;)
        db[grp][r] = assocs
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Loading 2024-11-03 human
Reading https://release.geneontology.org/2024-11-03/annotations/goa_human.gaf.gz using ParserConfiguration(preserve_negated_associations=True, include_association_attributes=None, primary_knowledge_source=None, aggregator_knowledge_source=None)
Loaded 782823 associations
  1494 negated associations
Loading 2024-11-03 Arabidopsis thaliana
Reading https://release.geneontology.org/2024-11-03/annotations/tair.gaf.gz using ParserConfiguration(preserve_negated_associations=True, include_association_attributes=None, primary_knowledge_source=None, aggregator_knowledge_source=None)
Loaded 235371 associations
  1374 negated associations
Loading 2024-11-03 yeast
Reading https://release.geneontology.org/2024-11-03/annotations/sgd.gaf.gz using ParserConfiguration(preserve_negated_associations=True, include_association_attributes=None, primary_knowledge_source=None, aggregator_knowledge_source=None)
Loaded 120823 associations
  6 negated associations
Loading 2024-06-10 human
Reading https://release.geneontology.org/2024-06-10/annotations/goa_human.gaf.gz using ParserConfiguration(preserve_negated_associations=True, include_association_attributes=None, primary_knowledge_source=None, aggregator_knowledge_source=None)
Loaded 707168 associations
  1308 negated associations
Loading 2024-06-10 Arabidopsis thaliana
Reading https://release.geneontology.org/2024-06-10/annotations/tair.gaf.gz using ParserConfiguration(preserve_negated_associations=True, include_association_attributes=None, primary_knowledge_source=None, aggregator_knowledge_source=None)
Loaded 235504 associations
  1373 negated associations
Loading 2024-06-10 yeast
Reading https://release.geneontology.org/2024-06-10/annotations/sgd.gaf.gz using ParserConfiguration(preserve_negated_associations=True, include_association_attributes=None, primary_knowledge_source=None, aggregator_knowledge_source=None)
Loaded 117290 associations
  7 negated associations
Loading 2020-01-01 human
Reading https://release.geneontology.org/2020-01-01/annotations/goa_human.gaf.gz using ParserConfiguration(preserve_negated_associations=True, include_association_attributes=None, primary_knowledge_source=None, aggregator_knowledge_source=None)
Loaded 495361 associations
  1244 negated associations
Loading 2020-01-01 Arabidopsis thaliana
Reading https://release.geneontology.org/2020-01-01/annotations/tair.gaf.gz using ParserConfiguration(preserve_negated_associations=True, include_association_attributes=None, primary_knowledge_source=None, aggregator_knowledge_source=None)
Loaded 236821 associations
  1364 negated associations
Loading 2020-01-01 yeast
Reading https://release.geneontology.org/2020-01-01/annotations/sgd.gaf.gz using ParserConfiguration(preserve_negated_associations=True, include_association_attributes=None, primary_knowledge_source=None, aggregator_knowledge_source=None)
Loaded 120916 associations
  28 negated associations
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>db[&quot;goa_human&quot;][LATEST][0]
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Association(subject=&#39;UniProtKB:A0A024RBG1&#39;, predicate=&#39;enables&#39;, object=&#39;GO:0003723&#39;, property_values=[], subject_label=&#39;NUDT4B&#39;, predicate_label=None, object_label=None, negated=None, publications=[&#39;GO_REF:0000043&#39;], evidence_type=&#39;IEA&#39;, supporting_objects=[], primary_knowledge_source=&#39;infores:UniProt&#39;, aggregator_knowledge_source=None, subject_closure=[], subject_closure_label=[], object_closure=[], object_closure_label=[], comments=[])
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>len([x for x in db[&quot;goa_human&quot;][LATEST] if x.negated])
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
1494
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span># reload modules
<br/></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>%load_ext autoreload
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[53]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>%autoreload 0
</pre></div>
</div>
</div>
</section>
<section id="Diffs-by-terms">
<h2>Diffs by terms<a class="headerlink" href="#Diffs-by-terms" title="Link to this heading"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>from oaklib.utilities.associations.association_differ import AssociationDiffer


differ = AssociationDiffer(adapter=go)
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>cache = {}
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[73]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>len(db[&quot;goa_human&quot;].keys())
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[73]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
3
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[74]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>list(db[&quot;goa_human&quot;].keys())
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[74]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;2024-11-03&#39;, &#39;2024-06-10&#39;, &#39;2020-01-01&#39;]
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[89]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>#ix = differ.changes_by_terms(db[&quot;goa_human&quot;][LATEST], db[&quot;goa_human&quot;][PREVIOUS[0]], min_num_entities_changes=10, cache={})
grp = &quot;sgd&quot;
ix = differ.changes_by_terms(db[grp][PREVIOUS[0]], db[grp][LATEST], min_num_entities_changes=2, cache={})
<br/></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[90]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>len(ix)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[90]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
847
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[91]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>for k in list(ix.keys())[0:5]:
    print(k, go.label(k))
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
GO:1904688 regulation of cytoplasmic translational initiation
GO:0170039 proteinogenic amino acid metabolic process
GO:0005980 glycogen catabolic process
GO:0045937 positive regulation of phosphate metabolic process
GO:0042762 regulation of sulfur metabolic process
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[92]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>pubmed_adapter = get_adapter(&quot;pubmed:&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
WARNING:eutils._internal.queryservice:No NCBI API key provided; throttling to 3 requests/second; see https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[93]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>from functools import lru_cache


@lru_cache
def pub_date(pmid):
    m = pubmed_adapter.entity_metadata_map(pmid)
    if m:
        return m.get(&quot;year&quot;)
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[110]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>from typing import Optional
from oaklib.utilities.associations.association_differ import TermComparison


def score_term_comparison(term: str, ix: TermComparison, max_date_inclusive=None, max_genes=20) -&gt; Optional[dict]:
    &quot;&quot;&quot;
    We are interested in genes for which all evidence was from previously known.

    :param term:
    :param ix:
    :return:
    &quot;&quot;&quot;
    #all_genes = set(ix.old_associations_by_entity.keys()).union(ix.new_associations_by_entity.keys())
    #gene_diff = all_genes - set(ix.old_associations_by_entity.keys()).intersection(ix.new_associations_by_entity.keys())
    if len(ix.new_associations_by_entity) &gt; max_genes:
        return None
    new_genes = set(ix.new_associations_by_entity.keys()) - set(ix.old_associations_by_entity.keys())
    gene_id_to_label_map = {}
    for g, assocs in ix.new_associations_by_entity.items():
        gene_id_to_label_map[g] = assocs[0].subject_label
    for g, assocs in ix.old_associations_by_entity.items():
        gene_id_to_label_map[g] = assocs[0].subject_label
    filtered_new_genes = {}
    for gene in new_genes:
        all_before_cutoff = True
        for a in ix.new_associations_by_entity[gene]:
            pmids = [x for x in a.publications if x.startswith(&quot;PMID&quot;)]
            if max_date_inclusive is not None:
                # print(f&quot;{term} {gene}, pmids={pmids}&quot;)
                if not pmids:
                    all_before_cutoff = False
                else:
                    pmid = pmids[0]
                    date = pub_date(pmid)
                    if date is None or date &gt; max_date_inclusive:
                        all_before_cutoff = False
        if all_before_cutoff:
            filtered_new_genes[gene] = ix.new_associations_by_entity[gene]
    #gene_id_diff = len(gene_diff)
    term_lbl = go.label(term)
    if term in mf_terms:
        ann_pred = &quot;that are capable of&quot;
    elif term in bp_terms:
        ann_pred = &quot;involved in&quot;
    elif term in cc_terms:
        ann_pred = &quot;localized to&quot;
    else:
        return
    def as_genes_list(amap):
        return [gene_id_to_label_map[g] for g in amap]
    def as_str_list(amap):
        return [str(g) for g in amap]
    case = {
        &quot;input&quot;: f&quot;List all genes {ann_pred} {term_lbl}&quot;,
        &quot;ideal&quot;: &quot;; &quot;.join(as_genes_list(ix.new_associations_by_entity)),
        &quot;original_input&quot;: {
            &quot;term&quot;: str(term),
            &quot;genes_current&quot;: as_str_list(ix.new_associations_by_entity),
            &quot;genes_previous&quot;: as_str_list(ix.old_associations_by_entity),
            &quot;genes_added&quot;: as_genes_list(new_genes),
            &quot;genes_added_prior_to_cutoff&quot;: as_genes_list(filtered_new_genes),
            &quot;num_genes_added_prior_to_cutoff&quot;: len(filtered_new_genes),
            &quot;date_cutoff&quot;: max_date_inclusive,
        }
    }
    return case
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[111]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>import yaml

n = 0
for k in list(ix.keys()):
    lbl = go.label(k)
    if &quot;regulation&quot; in lbl:
        continue
    if &quot;response to&quot; in lbl:
        continue
    case = score_term_comparison(k, ix[k], max_date_inclusive=&quot;2022&quot;)
    if not case:
        continue
    if case[&quot;original_input&quot;][&quot;num_genes_added_prior_to_cutoff&quot;] &gt; 2:
        print(yaml.dump(case, sort_keys=False))
        n += 1
    if n &gt; 40:
        break
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
input: List all genes localized to respiratory chain complex IV
ideal: COX6; COX12; COX5B; COX9; COX5A; COX4; COX8; MTC3; COX13; COX7; COX1; COX2;
  COX3; COX26; AI4; AI5_ALPHA; AI3
original_input:
  term: GO:0045277
  genes_current:
  - SGD:S000001093
  - SGD:S000004028
  - SGD:S000001373
  - SGD:S000002225
  - SGD:S000004997
  - SGD:S000003155
  - SGD:S000004387
  - SGD:S000003195
  - SGD:S000003159
  - SGD:S000004869
  - SGD:S000007260
  - SGD:S000007281
  - SGD:S000007283
  - SGD:S000113555
  - SGD:S000007264
  - SGD:S000007265
  - SGD:S000007263
  genes_previous:
  - SGD:S000004387
  - SGD:S000004028
  - SGD:S000004857
  - SGD:S000007260
  genes_added:
  - COX26
  - COX13
  - AI4
  - COX4
  - COX9
  - AI3
  - MTC3
  - COX6
  - COX3
  - AI5_ALPHA
  - COX5A
  - COX5B
  - COX7
  - COX2
  genes_added_prior_to_cutoff:
  - COX26
  - COX9
  - COX3
  - COX7
  - COX2
  num_genes_added_prior_to_cutoff: 5
  date_cutoff: &#39;2022&#39;

input: List all genes localized to TTT Hsp90 cochaperone complex
ideal: TTI2; RVB2; TRA1; TTI1; RVB1; ASA1; TEL2
original_input:
  term: GO:0110078
  genes_current:
  - SGD:S000003897
  - SGD:S000006156
  - SGD:S000001141
  - SGD:S000001516
  - SGD:S000002598
  - SGD:S000006289
  - SGD:S000003331
  genes_previous:
  - SGD:S000003897
  genes_added:
  - RVB1
  - TEL2
  - TTI1
  - ASA1
  - RVB2
  - TRA1
  genes_added_prior_to_cutoff:
  - RVB1
  - TEL2
  - TTI1
  - ASA1
  - RVB2
  - TRA1
  num_genes_added_prior_to_cutoff: 6
  date_cutoff: &#39;2022&#39;

input: List all genes localized to respiratory chain complex III
ideal: COR1; RIP1; QCR6; QCR8; QCR7; QCR9; QCR10; CYT1; QCR2; COB
original_input:
  term: GO:0045275
  genes_current:
  - SGD:S000000141
  - SGD:S000000750
  - SGD:S000001929
  - SGD:S000003702
  - SGD:S000002937
  - SGD:S000003415
  - SGD:S000003529
  - SGD:S000005591
  - SGD:S000006395
  - SGD:S000007270
  genes_previous:
  - SGD:S000007270
  genes_added:
  - QCR10
  - COR1
  - CYT1
  - QCR2
  - QCR9
  - QCR8
  - QCR6
  - QCR7
  - RIP1
  genes_added_prior_to_cutoff:
  - QCR10
  - COR1
  - QCR2
  num_genes_added_prior_to_cutoff: 3
  date_cutoff: &#39;2022&#39;

input: List all genes that are capable of alpha-1,4-glucosidase activity
ideal: MAL62; MAL42; MAL22; MAL32; GTB1; MAL12; IMA1; IMA2; IMA3; IMA4; IMA5
original_input:
  term: GO:0004558
  genes_current:
  - SGD:S000029690
  - SGD:S000029687
  - SGD:S000029682
  - SGD:S000000503
  - SGD:S000002629
  - SGD:S000003524
  - SGD:S000003519
  - SGD:S000005517
  - SGD:S000001434
  - SGD:S000003757
  - SGD:S000003752
  genes_previous:
  - SGD:S000002629
  genes_added:
  - MAL42
  - IMA1
  - IMA3
  - IMA4
  - IMA5
  - IMA2
  - MAL22
  - MAL32
  - MAL62
  - MAL12
  genes_added_prior_to_cutoff:
  - MAL42
  - MAL22
  - MAL62
  num_genes_added_prior_to_cutoff: 3
  date_cutoff: &#39;2022&#39;

input: List all genes that are capable of G-quadruplex DNA binding
ideal: RAP1; MGS1; SUB1; DNA2; NSR1; VID22; MSS116; XRS2; SLX9; PIF1; MRE11; DBP2;
  RAD50; DED1; DBP1; RRM3
original_input:
  term: GO:0051880
  genes_current:
  - SGD:S000005160
  - SGD:S000005162
  - SGD:S000004642
  - SGD:S000001207
  - SGD:S000003391
  - SGD:S000004365
  - SGD:S000002602
  - SGD:S000002777
  - SGD:S000003313
  - SGD:S000004526
  - SGD:S000004837
  - SGD:S000005056
  - SGD:S000005194
  - SGD:S000005730
  - SGD:S000006040
  - SGD:S000001073
  genes_previous:
  - SGD:S000004526
  - SGD:S000005160
  - SGD:S000005194
  - SGD:S000004837
  - SGD:S000002777
  - SGD:S000005162
  - SGD:S000004642
  - SGD:S000001207
  - SGD:S000003391
  - SGD:S000003313
  - SGD:S000001073
  - SGD:S000004365
  genes_added:
  - DBP2
  - MSS116
  - DED1
  - DBP1
  genes_added_prior_to_cutoff:
  - DBP2
  - MSS116
  - DED1
  - DBP1
  num_genes_added_prior_to_cutoff: 4
  date_cutoff: &#39;2022&#39;

input: List all genes that are capable of alpha-glucosidase activity
ideal: MAL62; MAL42; MAL22; ROT2; MAL32; SGA1; IMA4; IMA3; GTB1; CWH41; MAL12; IMA1;
  IMA5; GDB1; IMA2; YMR196W; STA1; CPX-417; SUC2
original_input:
  term: GO:0090599
  genes_current:
  - SGD:S000029690
  - SGD:S000029687
  - SGD:S000029682
  - SGD:S000000433
  - SGD:S000000503
  - SGD:S000001361
  - SGD:S000003757
  - SGD:S000001434
  - SGD:S000002629
  - SGD:S000002995
  - SGD:S000003524
  - SGD:S000003519
  - SGD:S000003752
  - SGD:S000006388
  - SGD:S000005517
  - SGD:S000004809
  - SGD:S000029522
  - SGD:S000217621
  - SGD:S000001424
  genes_previous:
  - SGD:S000001361
  - SGD:S000001434
  - SGD:S000003757
  - SGD:S000000433
  - SGD:S000000503
  - SGD:S000003752
  - SGD:S000002995
  - SGD:S000003519
  - SGD:S000003524
  - SGD:S000002629
  - SGD:S000006388
  - SGD:S000005517
  - SGD:S000004809
  - SGD:S000029522
  - SGD:S000217621
  - SGD:S000001424
  genes_added:
  - MAL42
  - MAL22
  - MAL62
  genes_added_prior_to_cutoff:
  - MAL42
  - MAL22
  - MAL62
  num_genes_added_prior_to_cutoff: 3
  date_cutoff: &#39;2022&#39;

</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>
<section id="OLD-ANALYSIS-BELOW">
<h2>OLD ANALYSIS BELOW<a class="headerlink" href="#OLD-ANALYSIS-BELOW" title="Link to this heading"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[289]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>from functools import lru_cache



@lru_cache
def lineage(t: str):
    up = set(go.ancestors(t, predicates=[IS_A, PART_OF], reflexive=True))
    dn = set(go.descendants(t, predicates=[IS_A, PART_OF]))
    return up.union(dn)

#len(lineage(&quot;GO:0005737&quot;))

@lru_cache
def ancs(t: str):
    return set(go.ancestors(t, predicates=[IS_A, PART_OF], reflexive=True))
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[290]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>import pandas as pd

def pmid(a):
    pubs = [p for p in a.publications if p.startswith(&quot;PMID&quot;)]
    if pubs:
        if len(pubs) &gt; 1:
            raise ValueError(f&quot;Multiple PMIDs: {pubs}&quot;)
        return pubs[0]
    return None

@lru_cache
def lbl(t: str):
    return go.label(t)

def assocs_to_df(assocs: list, release: str):

    df = pd.DataFrame([{
        &quot;subject&quot;: a.subject,
        &quot;subject_label&quot;: a.subject_label,
        &quot;predicate&quot;: a.predicate,
        &quot;object&quot;: a.object,
        &quot;object_label&quot;: lbl(a.object),
        &quot;object_obsoletes&quot;: a.object in obsoletes,
        &quot;object_uninformative&quot;: a.object in non_informative,
        &quot;object_closure&quot;: ancs(a.object),
        &quot;object_closure_redundant&quot;: ancs(a.object) - {a.object},
        &quot;evidence&quot;: a.evidence_type,
        &quot;is_iba&quot;: a.evidence_type == &quot;IBA&quot;,
        &quot;negated&quot;: a.negated,
        &quot;pmid&quot;: pmid(a),
        &quot;pubs&quot;: a.publications,
        &quot;release&quot;: release,
    } for a in assocs])
    return df
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[291]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>human_df = assocs_to_df(db[&quot;goa_human&quot;][LATEST], LATEST)
tair_df = assocs_to_df(db[&quot;tair&quot;][LATEST], LATEST)
sgd_df = assocs_to_df(db[&quot;sgd&quot;][LATEST], LATEST)
df = sgd_df
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[292]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>prev_df = assocs_to_df(db[&quot;sgd&quot;][PREVIOUS[-1]], PREVIOUS[-1])
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[293]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>df
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[293]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>subject</th>
      <th>subject_label</th>
      <th>predicate</th>
      <th>object</th>
      <th>object_label</th>
      <th>object_obsoletes</th>
      <th>object_uninformative</th>
      <th>object_closure</th>
      <th>object_closure_redundant</th>
      <th>evidence</th>
      <th>is_iba</th>
      <th>negated</th>
      <th>pmid</th>
      <th>pubs</th>
      <th>release</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>SGD:S000003381</td>
      <td>GPC1</td>
      <td>acts_upstream_of_or_within</td>
      <td>GO:0090640</td>
      <td>phosphatidylcholine biosynthesis from sn-glyce...</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0019637, GO:0008152, BFO:0000015, GO:00066...</td>
      <td>{GO:0019637, GO:0008152, GO:0006796, BFO:00000...</td>
      <td>IGI</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:30514764</td>
      <td>[PMID:30514764]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>1</th>
      <td>SGD:S000005701</td>
      <td>ALE1</td>
      <td>acts_upstream_of_or_within</td>
      <td>GO:0090640</td>
      <td>phosphatidylcholine biosynthesis from sn-glyce...</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0019637, GO:0008152, BFO:0000015, GO:00066...</td>
      <td>{GO:0019637, GO:0008152, GO:0006796, BFO:00000...</td>
      <td>IGI</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:30514764</td>
      <td>[PMID:30514764]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>2</th>
      <td>SGD:S000003381</td>
      <td>GPC1</td>
      <td>acts_upstream_of_or_within</td>
      <td>GO:0036151</td>
      <td>phosphatidylcholine acyl-chain remodeling</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0019637, GO:0008152, GO:0006796, BFO:00000...</td>
      <td>{GO:0019637, GO:0008152, GO:0006796, BFO:00000...</td>
      <td>IMP</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:30514764</td>
      <td>[PMID:30514764]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>3</th>
      <td>SGD:S000004492</td>
      <td>RCF1</td>
      <td>acts_upstream_of_or_within</td>
      <td>GO:0033617</td>
      <td>mitochondrial cytochrome c oxidase assembly</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0043933, GO:0044085, GO:0065003, GO:001604...</td>
      <td>{GO:0043933, GO:0044085, GO:0022607, GO:007184...</td>
      <td>IMP</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:29746825</td>
      <td>[PMID:29746825]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>4</th>
      <td>SGD:S000004977</td>
      <td>SIW14</td>
      <td>enables</td>
      <td>GO:0052845</td>
      <td>inositol-5-diphosphate-1,2,3,4,6-pentakisphosp...</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0016817, GO:0016818, GO:0052842, GO:000382...</td>
      <td>{GO:0016817, GO:0016818, GO:0052842, GO:000382...</td>
      <td>IDA</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:26828065</td>
      <td>[PMID:26828065]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>120818</th>
      <td>SGD:S000003241</td>
      <td>SEC9</td>
      <td>involved_in</td>
      <td>GO:0006906</td>
      <td>vesicle fusion</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0048284, GO:0051234, GO:0090174, GO:001604...</td>
      <td>{GO:0061024, BFO:0000015, GO:0009987, GO:00160...</td>
      <td>IBA</td>
      <td>True</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>120819</th>
      <td>SGD:S000004826</td>
      <td>CEF1</td>
      <td>part_of</td>
      <td>GO:0000974</td>
      <td>Prp19 complex</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0000974, BFO:0000004, BFO:0000040, BFO:000...</td>
      <td>{BFO:0000004, BFO:0000040, BFO:0000002, GO:003...</td>
      <td>IBA</td>
      <td>True</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>120820</th>
      <td>SGD:S000002551</td>
      <td>MKC7</td>
      <td>involved_in</td>
      <td>GO:0031505</td>
      <td>fungal-type cell wall organization</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0071554, GO:0016043, GO:0045229, GO:003150...</td>
      <td>{GO:0071554, GO:0016043, GO:0045229, GO:000998...</td>
      <td>IBA</td>
      <td>True</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>120821</th>
      <td>SGD:S000003008</td>
      <td>HEM2</td>
      <td>is_active_in</td>
      <td>GO:0005829</td>
      <td>cytosol</td>
      <td>False</td>
      <td>False</td>
      <td>{CARO:0030000, UBERON:0000061, CARO:0000003, G...</td>
      <td>{CARO:0030000, CARO:0000000, BFO:0000004, GO:0...</td>
      <td>IBA</td>
      <td>True</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>120822</th>
      <td>SGD:S000001122</td>
      <td>LAM4</td>
      <td>involved_in</td>
      <td>GO:0032366</td>
      <td>intracellular sterol transport</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0006869, GO:0015850, GO:0032365, GO:005123...</td>
      <td>{GO:0032365, GO:0015918, GO:0051649, GO:000998...</td>
      <td>IBA</td>
      <td>True</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
    </tr>
  </tbody>
</table>
<p>120823 rows × 15 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[294]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>from typing import Dict, List, Optional


def repair_assocs_df(assocs: pd.DataFrame):
    &quot;&quot;&quot;
    Ensures that IDs are normalized.

    :param assocs:
    :return:
    &quot;&quot;&quot;
    # may not be 1:1
    subject_label_to_ids: Dict[str, List[str]]
    subject_label_to_ids = assocs.groupby(&quot;subject_label&quot;)[&quot;subject&quot;].aggregate(lambda x: list(set(x))).to_dict()
    labels_with_multiple_ids = {k: v for k, v in subject_label_to_ids.items() if len(v) &gt; 1}
    if labels_with_multiple_ids:
        print(f&quot;Multiple IDs for {len(labels_with_multiple_ids)} labels&quot;)
        print(list(labels_with_multiple_ids.items())[:5])
    labels_to_canonical = {k: sorted(list(v))[0] for k, v in subject_label_to_ids.items()}
    assocs[&#39;subject&#39;] = assocs[&#39;subject_label&#39;].map(labels_to_canonical)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[295]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>test_df = tair_df.copy()
repair_assocs_df(test_df)
test_df
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Multiple IDs for 5258 labels
[(&#39;4CL1&#39;, [&#39;TAIR:locus:2017602&#39;, &#39;AGI_LocusCode:AT1G51680&#39;]), (&#39;4CL2&#39;, [&#39;TAIR:locus:2094716&#39;, &#39;AGI_LocusCode:AT3G21240&#39;]), (&#39;4CL3&#39;, [&#39;TAIR:locus:2015003&#39;, &#39;AGI_LocusCode:AT1G65060&#39;]), (&#39;AAC1&#39;, [&#39;TAIR:locus:2077778&#39;, &#39;AGI_LocusCode:AT3G08580&#39;]), (&#39;AAC2&#39;, [&#39;AGI_LocusCode:AT5G13490&#39;, &#39;TAIR:locus:2185041&#39;])]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[295]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>subject</th>
      <th>subject_label</th>
      <th>predicate</th>
      <th>object</th>
      <th>object_label</th>
      <th>object_obsoletes</th>
      <th>object_uninformative</th>
      <th>object_closure</th>
      <th>object_closure_redundant</th>
      <th>evidence</th>
      <th>is_iba</th>
      <th>negated</th>
      <th>pmid</th>
      <th>pubs</th>
      <th>release</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AGI_LocusCode:AT1G11880</td>
      <td>AT1G11880</td>
      <td>enables</td>
      <td>GO:0000009</td>
      <td>alpha-1,6-mannosyltransferase activity</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0000030, GO:0003824, GO:0016740, BFO:00000...</td>
      <td>{GO:0000030, GO:0003824, GO:0016740, BFO:00000...</td>
      <td>IEA</td>
      <td>False</td>
      <td>None</td>
      <td>None</td>
      <td>[TAIR:AnalysisReference:501756966]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AGI_LocusCode:AT1G80420</td>
      <td>ATXRCC1</td>
      <td>involved_in</td>
      <td>GO:0000012</td>
      <td>single strand break repair</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0043170, GO:0033554, GO:0008152, GO:000613...</td>
      <td>{GO:0043170, GO:0008152, BFO:0000015, GO:00099...</td>
      <td>IEA</td>
      <td>False</td>
      <td>None</td>
      <td>None</td>
      <td>[TAIR:AnalysisReference:501756966]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AGI_LocusCode:AT1G74030</td>
      <td>ENO1</td>
      <td>part_of</td>
      <td>GO:0000015</td>
      <td>phosphopyruvate hydratase complex</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0005829, GO:0110165, BFO:0000002, GO:00329...</td>
      <td>{CARO:0030000, GO:0005829, BFO:0000004, CARO:0...</td>
      <td>IEA</td>
      <td>False</td>
      <td>None</td>
      <td>None</td>
      <td>[TAIR:AnalysisReference:501756966]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AGI_LocusCode:AT2G29560</td>
      <td>ENOC</td>
      <td>part_of</td>
      <td>GO:0000015</td>
      <td>phosphopyruvate hydratase complex</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0005829, GO:0110165, BFO:0000002, GO:00329...</td>
      <td>{CARO:0030000, GO:0005829, BFO:0000004, CARO:0...</td>
      <td>IEA</td>
      <td>False</td>
      <td>None</td>
      <td>None</td>
      <td>[TAIR:AnalysisReference:501756966]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AGI_LocusCode:AT2G36530</td>
      <td>LOS2</td>
      <td>part_of</td>
      <td>GO:0000015</td>
      <td>phosphopyruvate hydratase complex</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0005829, GO:0110165, BFO:0000002, GO:00329...</td>
      <td>{CARO:0030000, GO:0005829, BFO:0000004, CARO:0...</td>
      <td>IEA</td>
      <td>False</td>
      <td>None</td>
      <td>None</td>
      <td>[TAIR:AnalysisReference:501756966]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>235366</th>
      <td>TAIR:locus:2058630</td>
      <td>At2g23210</td>
      <td>enables</td>
      <td>GO:0010294</td>
      <td>abscisic acid glucosyltransferase activity</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0035251, GO:0003824, GO:0016740, BFO:00000...</td>
      <td>{GO:0035251, GO:0003824, GO:0016740, BFO:00000...</td>
      <td>IBA</td>
      <td>True</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>235367</th>
      <td>AGI_LocusCode:AT2G15820</td>
      <td>OTP51</td>
      <td>involved_in</td>
      <td>GO:0045292</td>
      <td>mRNA cis splicing, via spliceosome</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0006397, GO:0008152, GO:0000375, BFO:00000...</td>
      <td>{GO:0009059, GO:0043170, GO:0006397, GO:000815...</td>
      <td>IBA</td>
      <td>True</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>235368</th>
      <td>TAIR:locus:2143196</td>
      <td>At5g15750</td>
      <td>involved_in</td>
      <td>GO:0042274</td>
      <td>ribosomal small subunit biogenesis</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0042274, GO:0044085, GO:0009987, BFO:00000...</td>
      <td>{GO:0044085, GO:0009987, BFO:0000015, GO:00226...</td>
      <td>IBA</td>
      <td>True</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>235369</th>
      <td>AGI_LocusCode:AT4G14730</td>
      <td>LFG1</td>
      <td>is_active_in</td>
      <td>GO:0016020</td>
      <td>membrane</td>
      <td>False</td>
      <td>False</td>
      <td>{CARO:0030000, UBERON:0000061, CARO:0000003, G...</td>
      <td>{CARO:0030000, UBERON:0000061, CARO:0000003, G...</td>
      <td>IBA</td>
      <td>True</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>235370</th>
      <td>TAIR:locus:2116525</td>
      <td>SD25</td>
      <td>enables</td>
      <td>GO:0004672</td>
      <td>protein kinase activity</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0140096, GO:0003824, GO:0016740, BFO:00000...</td>
      <td>{GO:0140096, GO:0003824, GO:0016740, BFO:00000...</td>
      <td>IBA</td>
      <td>True</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
    </tr>
  </tbody>
</table>
<p>235371 rows × 15 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[296]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>test_df[test_df[&#39;subject_label&#39;] == &quot;GALT6&quot;]
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[296]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>subject</th>
      <th>subject_label</th>
      <th>predicate</th>
      <th>object</th>
      <th>object_label</th>
      <th>object_obsoletes</th>
      <th>object_uninformative</th>
      <th>object_closure</th>
      <th>object_closure_redundant</th>
      <th>evidence</th>
      <th>is_iba</th>
      <th>negated</th>
      <th>pmid</th>
      <th>pubs</th>
      <th>release</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>56871</th>
      <td>AGI_LocusCode:AT5G62620</td>
      <td>GALT6</td>
      <td>located_in</td>
      <td>GO:0005794</td>
      <td>Golgi apparatus</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0110165, GO:0043231, BFO:0000002, GO:00057...</td>
      <td>{CARO:0030000, CARO:0000000, BFO:0000004, GO:0...</td>
      <td>ISM</td>
      <td>False</td>
      <td>None</td>
      <td>None</td>
      <td>[TAIR:AnalysisReference:501780126]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>56872</th>
      <td>AGI_LocusCode:AT5G62620</td>
      <td>GALT6</td>
      <td>located_in</td>
      <td>GO:0005794</td>
      <td>Golgi apparatus</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0110165, GO:0043231, BFO:0000002, GO:00057...</td>
      <td>{CARO:0030000, CARO:0000000, BFO:0000004, GO:0...</td>
      <td>ISM</td>
      <td>False</td>
      <td>None</td>
      <td>None</td>
      <td>[TAIR:AnalysisReference:501780126]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>61144</th>
      <td>AGI_LocusCode:AT5G62620</td>
      <td>GALT6</td>
      <td>involved_in</td>
      <td>GO:0006486</td>
      <td>protein glycosylation</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0008152, BFO:0000015, GO:0009100, GO:00434...</td>
      <td>{GO:0009059, GO:0043170, GO:0070085, GO:000815...</td>
      <td>IEA</td>
      <td>False</td>
      <td>None</td>
      <td>None</td>
      <td>[TAIR:AnalysisReference:501757242]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>85487</th>
      <td>AGI_LocusCode:AT5G62620</td>
      <td>GALT6</td>
      <td>enables</td>
      <td>GO:0030246</td>
      <td>carbohydrate binding</td>
      <td>False</td>
      <td>True</td>
      <td>{BFO:0000015, GO:0003674, GO:0005488, BFO:0000...</td>
      <td>{GO:0003674, BFO:0000015, GO:0005488, BFO:0000...</td>
      <td>IEA</td>
      <td>False</td>
      <td>None</td>
      <td>None</td>
      <td>[TAIR:AnalysisReference:501756966]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>165846</th>
      <td>AGI_LocusCode:AT5G62620</td>
      <td>GALT6</td>
      <td>located_in</td>
      <td>GO:0005794</td>
      <td>Golgi apparatus</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0110165, GO:0043231, BFO:0000002, GO:00057...</td>
      <td>{CARO:0030000, CARO:0000000, BFO:0000004, GO:0...</td>
      <td>IDA</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:26690932</td>
      <td>[TAIR:Publication:501767599, PMID:26690932]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>165851</th>
      <td>AGI_LocusCode:AT5G62620</td>
      <td>GALT6</td>
      <td>involved_in</td>
      <td>GO:0010405</td>
      <td>arabinogalactan protein metabolic process</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0043170, GO:0044036, GO:0010384, GO:007155...</td>
      <td>{GO:0043170, GO:0044036, GO:0071554, GO:000815...</td>
      <td>IMP</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:26690932</td>
      <td>[TAIR:Publication:501767599, PMID:26690932]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>165856</th>
      <td>AGI_LocusCode:AT5G62620</td>
      <td>GALT6</td>
      <td>involved_in</td>
      <td>GO:0018258</td>
      <td>protein O-linked glycosylation via hydroxyproline</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0006493, GO:0008152, BFO:0000015, GO:00091...</td>
      <td>{GO:0009059, GO:0006493, GO:0043170, GO:007008...</td>
      <td>IDA</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:26690932</td>
      <td>[TAIR:Publication:501767599, PMID:26690932]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>165858</th>
      <td>AGI_LocusCode:AT5G62620</td>
      <td>GALT6</td>
      <td>acts_upstream_of_or_within</td>
      <td>GO:0048354</td>
      <td>mucilage biosynthetic process involved in seed...</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0032501, GO:0008152, GO:0048359, BFO:00000...</td>
      <td>{GO:0032501, GO:0010192, GO:0008152, GO:004835...</td>
      <td>IMP</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:26690932</td>
      <td>[TAIR:Publication:501767599, PMID:26690932]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>165863</th>
      <td>AGI_LocusCode:AT5G62620</td>
      <td>GALT6</td>
      <td>acts_upstream_of_or_within</td>
      <td>GO:1900056</td>
      <td>negative regulation of leaf senescence</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0065007, BFO:0000015, GO:1900055, GO:00485...</td>
      <td>{GO:0065007, BFO:0000015, GO:1900055, GO:00485...</td>
      <td>IMP</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:26690932</td>
      <td>[TAIR:Publication:501767599, PMID:26690932]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>165869</th>
      <td>AGI_LocusCode:AT5G62620</td>
      <td>GALT6</td>
      <td>enables</td>
      <td>GO:1990714</td>
      <td>hydroxyproline O-galactosyltransferase activity</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0003824, GO:0016740, BFO:0000015, GO:00083...</td>
      <td>{GO:0003824, GO:0016740, BFO:0000015, GO:00083...</td>
      <td>IDA</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:26690932</td>
      <td>[TAIR:Publication:501767599, PMID:26690932]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>210076</th>
      <td>AGI_LocusCode:AT5G62620</td>
      <td>GALT6</td>
      <td>enables</td>
      <td>GO:1990714</td>
      <td>hydroxyproline O-galactosyltransferase activity</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0003824, GO:0016740, BFO:0000015, GO:00083...</td>
      <td>{GO:0003824, GO:0016740, BFO:0000015, GO:00083...</td>
      <td>IBA</td>
      <td>True</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>217996</th>
      <td>AGI_LocusCode:AT5G62620</td>
      <td>GALT6</td>
      <td>is_active_in</td>
      <td>GO:0000139</td>
      <td>Golgi membrane</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0110165, GO:0043231, BFO:0000002, GO:00057...</td>
      <td>{CARO:0030000, GO:0005794, CARO:0000000, BFO:0...</td>
      <td>IBA</td>
      <td>True</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[297]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><br/><span></span>def create_gene_df(df: pd.DataFrame):
    &quot;&quot;&quot;
    creates a new dataframe, grouped by gene (subject)

    :param df:
    :return:
    &quot;&quot;&quot;
    gene_df = df.groupby(&quot;subject&quot;).agg({
        &quot;object&quot;: &quot;count&quot;,
        &quot;is_iba&quot;: &quot;sum&quot;,
        &quot;negated&quot;: &quot;sum&quot;,
        # for object, take the union of all distinct values
        &quot;object&quot;: lambda x: set(x),
        # for object closure, take the union of all sets
        &quot;object_closure&quot;: lambda x: set.union(*x),
        &quot;object_closure_redundant&quot;: lambda x: set.union(*x),
        &quot;pmid&quot;: lambda x: set(x),
        #&quot;pubs&quot;: lambda x: set.union(set(x)),
        &quot;release&quot;: &quot;first&quot;,
    }).reset_index()
    # the redundant closure is the set difference of object_closure and object
    gene_df = gene_df.rename(columns={
        &quot;object&quot;: &quot;terms&quot;,
        &quot;is_iba&quot;: &quot;n_iba&quot;,
        &quot;negated&quot;: &quot;n_negated&quot;,
        &quot;object_closure&quot;: &quot;closure&quot;,
        &quot;pmid&quot;: &quot;n_pmid&quot;,
    })
    return gene_df
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[298]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>gene_df = create_gene_df(df)
gene_df
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[298]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>subject</th>
      <th>terms</th>
      <th>n_iba</th>
      <th>n_negated</th>
      <th>closure</th>
      <th>object_closure_redundant</th>
      <th>n_pmid</th>
      <th>release</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>SGD:S000000001</td>
      <td>{GO:0071168, GO:0005739, GO:0008301, GO:000367...</td>
      <td>3</td>
      <td>0</td>
      <td>{GO:0005739, GO:0000182, GO:0008152, BFO:00000...</td>
      <td>{GO:0000182, GO:0008152, BFO:0000015, GO:00010...</td>
      <td>{PMID:2404611, PMID:18708580, PMID:2649882, PM...</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>1</th>
      <td>SGD:S000000002</td>
      <td>{GO:0015031, GO:0099023, GO:0005768, GO:000662...</td>
      <td>4</td>
      <td>0</td>
      <td>{GO:0032509, GO:0071985, GO:0046872, BFO:00000...</td>
      <td>{GO:0032509, GO:0071985, GO:0046872, BFO:00000...</td>
      <td>{PMID:19828734, PMID:30358795, PMID:20173035, ...</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>2</th>
      <td>SGD:S000000003</td>
      <td>{GO:0005085, GO:0005737, GO:0032232, GO:000582...</td>
      <td>3</td>
      <td>0</td>
      <td>{GO:0005085, GO:0030234, GO:0008152, BFO:00000...</td>
      <td>{GO:0030234, GO:0008152, BFO:0000015, GO:00900...</td>
      <td>{PMID:19545407, PMID:10409717, PMID:17925388, ...</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>3</th>
      <td>SGD:S000000004</td>
      <td>{GO:0005829, GO:0072671, GO:0005576, GO:003460...</td>
      <td>9</td>
      <td>0</td>
      <td>{GO:1901363, GO:0003723, GO:0008152, BFO:00000...</td>
      <td>{GO:1901363, GO:0003723, GO:0008152, BFO:00000...</td>
      <td>{PMID:18706386, PMID:9789005, PMID:26928762, P...</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>4</th>
      <td>SGD:S000000005</td>
      <td>{GO:0015031, GO:0006621, GO:0005789, GO:000688...</td>
      <td>8</td>
      <td>0</td>
      <td>{BFO:0000015, GO:0030135, GO:0070972, GO:01101...</td>
      <td>{BFO:0000015, GO:0030135, GO:0070972, GO:01101...</td>
      <td>{PMID:26928762, PMID:10359606, None, PMID:1115...</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>6906</th>
      <td>SGD:S000350095</td>
      <td>{GO:0008150, GO:0003674, GO:0005575}</td>
      <td>0</td>
      <td>0</td>
      <td>{BFO:0000003, BFO:0000015, BFO:0000004, GO:000...</td>
      <td>{BFO:0000004, BFO:0000040, BFO:0000003, BFO:00...</td>
      <td>{None}</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>6907</th>
      <td>SGD:S000350096</td>
      <td>{GO:0008150, GO:0003674, GO:0005575}</td>
      <td>0</td>
      <td>0</td>
      <td>{BFO:0000003, BFO:0000015, BFO:0000004, GO:000...</td>
      <td>{BFO:0000004, BFO:0000040, BFO:0000003, BFO:00...</td>
      <td>{None}</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>6908</th>
      <td>SGD:S000350097</td>
      <td>{GO:0005575, GO:0003674, GO:0008150}</td>
      <td>0</td>
      <td>0</td>
      <td>{BFO:0000004, BFO:0000002, GO:0003674, GO:0008...</td>
      <td>{BFO:0000004, BFO:0000002, BFO:0000003, BFO:00...</td>
      <td>{None}</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>6909</th>
      <td>SGD:S000350098</td>
      <td>{GO:0005575, GO:0003674, GO:0008150}</td>
      <td>0</td>
      <td>0</td>
      <td>{BFO:0000004, BFO:0000002, GO:0003674, GO:0008...</td>
      <td>{BFO:0000004, BFO:0000002, BFO:0000003, BFO:00...</td>
      <td>{None}</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>6910</th>
      <td>SGD:S000350099</td>
      <td>{GO:0003674, GO:0005575, GO:0008150}</td>
      <td>0</td>
      <td>0</td>
      <td>{BFO:0000040, BFO:0000015, BFO:0000004, GO:000...</td>
      <td>{BFO:0000004, BFO:0000002, BFO:0000003, BFO:00...</td>
      <td>{None}</td>
      <td>2024-11-03</td>
    </tr>
  </tbody>
</table>
<p>6911 rows × 8 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[299]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>def set_redundant_flag(assocs: pd.DataFrame, gene_df: pd.DataFrame):
    &quot;&quot;&quot;
    Sets the redundant flag for each association, if the object is in object_closure_redundant for that gene.

    :param assocs:
    :param gene_df:
    :return:
    &quot;&quot;&quot;
    # Create a mapping of subject to object_closure_redundant
    redundant_map = gene_df.set_index(&#39;subject&#39;)[&#39;object_closure_redundant&#39;].to_dict()

    # Vectorized check for each row
    def check_redundant(row):
        if row[&#39;is_iba&#39;]:
            return False
        closure_set = redundant_map.get(row[&#39;subject&#39;], set())
        is_redundant = row[&#39;object&#39;] in closure_set
        return is_redundant

    # Apply the check to all rows at once
    assocs[&#39;redundant&#39;] = assocs.apply(check_redundant, axis=1)
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[300]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>df[&quot;redundant&quot;] = False
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[301]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>set_redundant_flag(df, gene_df)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[302]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>df[df[&#39;redundant&#39;]]
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[302]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>subject</th>
      <th>subject_label</th>
      <th>predicate</th>
      <th>object</th>
      <th>object_label</th>
      <th>object_obsoletes</th>
      <th>object_uninformative</th>
      <th>object_closure</th>
      <th>object_closure_redundant</th>
      <th>evidence</th>
      <th>is_iba</th>
      <th>negated</th>
      <th>pmid</th>
      <th>pubs</th>
      <th>release</th>
      <th>redundant</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>16</th>
      <td>SGD:S000004539</td>
      <td>FPR3</td>
      <td>located_in</td>
      <td>GO:0005634</td>
      <td>nucleus</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0110165, GO:0043231, BFO:0000002, GO:00432...</td>
      <td>{CARO:0030000, CARO:0000000, BFO:0000004, GO:0...</td>
      <td>IDA</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:26359986</td>
      <td>[PMID:26359986]</td>
      <td>2024-11-03</td>
      <td>True</td>
    </tr>
    <tr>
      <th>28</th>
      <td>SGD:S000002699</td>
      <td>HRQ1</td>
      <td>enables</td>
      <td>GO:0043138</td>
      <td>3'-5' DNA helicase activity</td>
      <td>False</td>
      <td>False</td>
      <td>{BFO:0000015, GO:0008150, GO:0016043, GO:00431...</td>
      <td>{GO:0140097, GO:0032508, GO:0003824, BFO:00000...</td>
      <td>IDA</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:28385527</td>
      <td>[PMID:28385527]</td>
      <td>2024-11-03</td>
      <td>True</td>
    </tr>
    <tr>
      <th>33</th>
      <td>SGD:S000002699</td>
      <td>HRQ1</td>
      <td>enables</td>
      <td>GO:0043138</td>
      <td>3'-5' DNA helicase activity</td>
      <td>False</td>
      <td>False</td>
      <td>{BFO:0000015, GO:0008150, GO:0016043, GO:00431...</td>
      <td>{GO:0140097, GO:0032508, GO:0003824, BFO:00000...</td>
      <td>IDA</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:24440721</td>
      <td>[PMID:24440721]</td>
      <td>2024-11-03</td>
      <td>True</td>
    </tr>
    <tr>
      <th>39</th>
      <td>SGD:S000003245</td>
      <td>SNU71</td>
      <td>located_in</td>
      <td>GO:0005634</td>
      <td>nucleus</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0110165, GO:0043231, BFO:0000002, GO:00432...</td>
      <td>{CARO:0030000, CARO:0000000, BFO:0000004, GO:0...</td>
      <td>IEA</td>
      <td>False</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000043]</td>
      <td>2024-11-03</td>
      <td>True</td>
    </tr>
    <tr>
      <th>42</th>
      <td>SGD:S000001443</td>
      <td>DJP1</td>
      <td>acts_upstream_of_or_within</td>
      <td>GO:0006626</td>
      <td>protein targeting to mitochondrion</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0070585, GO:0051234, GO:0006605, GO:007072...</td>
      <td>{GO:0070585, GO:0070727, GO:0033365, BFO:00000...</td>
      <td>IMP</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:30213914</td>
      <td>[PMID:30213914]</td>
      <td>2024-11-03</td>
      <td>True</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>105624</th>
      <td>SGD:S000006483</td>
      <td>RDN18-2</td>
      <td>located_in</td>
      <td>GO:0005840</td>
      <td>ribosome</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0110165, BFO:0000002, GO:0043229, CL:00000...</td>
      <td>{CARO:0030000, CARO:0000000, BFO:0000004, GO:0...</td>
      <td>IEA</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:30502926</td>
      <td>[PMID:30502926]</td>
      <td>2024-11-03</td>
      <td>True</td>
    </tr>
    <tr>
      <th>105625</th>
      <td>SGD:S000006502</td>
      <td>SNR42</td>
      <td>located_in</td>
      <td>GO:0005730</td>
      <td>nucleolus</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0043233, GO:0043231, GO:0110165, BFO:00000...</td>
      <td>{CARO:0030000, GO:0005634, CARO:0000000, BFO:0...</td>
      <td>IEA</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:30502926</td>
      <td>[PMID:30502926]</td>
      <td>2024-11-03</td>
      <td>True</td>
    </tr>
    <tr>
      <th>105627</th>
      <td>SGD:S000007300</td>
      <td>SNR36</td>
      <td>located_in</td>
      <td>GO:0005730</td>
      <td>nucleolus</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0043233, GO:0043231, GO:0110165, BFO:00000...</td>
      <td>{CARO:0030000, GO:0005634, CARO:0000000, BFO:0...</td>
      <td>IEA</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:30502926</td>
      <td>[PMID:30502926]</td>
      <td>2024-11-03</td>
      <td>True</td>
    </tr>
    <tr>
      <th>105628</th>
      <td>SGD:S000006484</td>
      <td>RDN25-1</td>
      <td>located_in</td>
      <td>GO:0005840</td>
      <td>ribosome</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0110165, BFO:0000002, GO:0043229, CL:00000...</td>
      <td>{CARO:0030000, CARO:0000000, BFO:0000004, GO:0...</td>
      <td>IEA</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:30502926</td>
      <td>[PMID:30502926]</td>
      <td>2024-11-03</td>
      <td>True</td>
    </tr>
    <tr>
      <th>105629</th>
      <td>SGD:S000006485</td>
      <td>RDN25-2</td>
      <td>located_in</td>
      <td>GO:0005840</td>
      <td>ribosome</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0110165, BFO:0000002, GO:0043229, CL:00000...</td>
      <td>{CARO:0030000, CARO:0000000, BFO:0000004, GO:0...</td>
      <td>IEA</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:30502926</td>
      <td>[PMID:30502926]</td>
      <td>2024-11-03</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
<p>34441 rows × 16 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[303]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>def annotate_new_pubs(assocs: pd.DataFrame, prev_assocs: pd.DataFrame):
    &quot;&quot;&quot;
    Annotates each association in assocs with fresh if its pmid is not in prev_assocs.

    :param assocs
    &quot;&quot;&quot;
    prev_pmid_map = prev_assocs.set_index(&#39;pmid&#39;)[&#39;subject&#39;].to_dict()

    # Vectorized check for each row
    def check_new(row):
        return row[&#39;pmid&#39;] not in prev_pmid_map

    # Apply the check to all rows in prev_assocs
    assocs[&#39;pmid_new&#39;] = assocs.apply(check_new, axis=1)

    # annotate tuple as unique if (subject, object) is not in prev
    prev_pair_assocs = set(zip(prev_assocs[&#39;subject&#39;], prev_assocs[&#39;object&#39;]))
    assocs[&#39;is_new&#39;] = ~assocs.apply(lambda x: (x[&#39;subject&#39;], x[&#39;object&#39;]) in prev_pair_assocs, axis=1)

    assocs[&#39;fresh&#39;] = assocs[&#39;pmid_new&#39;] &amp; assocs[&#39;is_new&#39;]
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[304]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>def annotate_redacted_pubs(assocs: pd.DataFrame, prev_assocs: pd.DataFrame):
    &quot;&quot;&quot;
    Annotates each association in prev_assocs with retracted if its pmid is not in assocs (latest).

    :param assocs
    &quot;&quot;&quot;
    # Create a mapping of pmid to subject
    pmid_map = assocs.set_index(&#39;pmid&#39;)[&#39;subject&#39;].to_dict()

    # Vectorized check for each row
    def check_removed(row):
        return row[&#39;pmid&#39;] not in pmid_map

    # Apply the check to all rows in prev_assocs
    prev_assocs[&#39;pmid_removed&#39;] = prev_assocs.apply(check_removed, axis=1)

    # annotate tuple as unique if (subject, object) is not in latest
    pair_assocs = set(zip(assocs[&#39;subject&#39;], assocs[&#39;object&#39;]))
    prev_assocs[&#39;unique&#39;] = ~prev_assocs.apply(lambda x: (x[&#39;subject&#39;], x[&#39;object&#39;]) in pair_assocs, axis=1)

    prev_assocs[&#39;redacted&#39;] = prev_assocs[&#39;pmid_removed&#39;] &amp; prev_assocs[&#39;unique&#39;]
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[305]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>annotate_redacted_pubs(df, prev_df)
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[306]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>#prev_df[&#39;redacted&#39;] = prev_df[&#39;pmid_removed&#39;] &amp; prev_df[&#39;unique&#39;]
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[307]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>#list(set(df[&#39;subject&#39;]))[:5]
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[308]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>#iba_df = df[df[&#39;is_iba&#39;]]
#iba_subjects = set(iba_df[&#39;subject&#39;])
#list(iba_subjects)[:5]
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[309]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>def set_iba_status(assocs: pd.DataFrame):
    iba_df = assocs[assocs[&#39;is_iba&#39;]]
    iba_subjects = set(iba_df[&#39;subject&#39;])
    # set the IBA rejected as associations for which
    # (a) the subject has an IBA association
    # (b) the object is not in the closure of the object in the IBA association
    # first we set the iba closure for each subject
    iba_closure = iba_df.groupby(&#39;subject&#39;)[&#39;object_closure&#39;].aggregate(lambda x: set.union(*x)).to_dict()
    # now we set IBA rejected for each row
    def check_iba_rejected(row):
        if row[&#39;is_iba&#39;]:
            return False
        return row[&#39;subject&#39;] in iba_subjects and row[&#39;object&#39;] not in iba_closure.get(row[&#39;subject&#39;], set())
        #if row[&#39;subject&#39;].startswith(&quot;TAIR&quot;):
        #    print(f&quot;Checking {row[&#39;subject&#39;]} in {list(iba_subjects)[:5]}&quot;)
        #return row[&#39;subject&#39;] in iba_subjects
        #return True

    assocs[&#39;iba_rejected&#39;] = assocs.apply(check_iba_rejected, axis=1)

#set_iba_status(df)
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[310]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>human_df
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[310]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>subject</th>
      <th>subject_label</th>
      <th>predicate</th>
      <th>object</th>
      <th>object_label</th>
      <th>object_obsoletes</th>
      <th>object_uninformative</th>
      <th>object_closure</th>
      <th>object_closure_redundant</th>
      <th>evidence</th>
      <th>is_iba</th>
      <th>negated</th>
      <th>pmid</th>
      <th>pubs</th>
      <th>release</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>UniProtKB:A0A024RBG1</td>
      <td>NUDT4B</td>
      <td>enables</td>
      <td>GO:0003723</td>
      <td>RNA binding</td>
      <td>False</td>
      <td>True</td>
      <td>{GO:0097159, GO:0003723, BFO:0000015, GO:00036...</td>
      <td>{GO:0097159, BFO:0000015, GO:0003674, GO:00036...</td>
      <td>IEA</td>
      <td>False</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000043]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>1</th>
      <td>UniProtKB:A0A024RBG1</td>
      <td>NUDT4B</td>
      <td>enables</td>
      <td>GO:0005515</td>
      <td>protein binding</td>
      <td>False</td>
      <td>True</td>
      <td>{BFO:0000015, GO:0003674, GO:0005488, BFO:0000...</td>
      <td>{GO:0003674, BFO:0000015, GO:0005488, BFO:0000...</td>
      <td>IPI</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:33961781</td>
      <td>[PMID:33961781]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>2</th>
      <td>UniProtKB:A0A024RBG1</td>
      <td>NUDT4B</td>
      <td>enables</td>
      <td>GO:0046872</td>
      <td>metal ion binding</td>
      <td>False</td>
      <td>True</td>
      <td>{GO:0043169, GO:0046872, BFO:0000015, GO:00360...</td>
      <td>{GO:0043169, BFO:0000015, GO:0036094, GO:00431...</td>
      <td>IEA</td>
      <td>False</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000043]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>3</th>
      <td>UniProtKB:A0A024RBG1</td>
      <td>NUDT4B</td>
      <td>located_in</td>
      <td>GO:0005829</td>
      <td>cytosol</td>
      <td>False</td>
      <td>False</td>
      <td>{CARO:0030000, UBERON:0000061, CARO:0000003, G...</td>
      <td>{CARO:0030000, CARO:0000000, BFO:0000004, GO:0...</td>
      <td>IDA</td>
      <td>False</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000052]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>4</th>
      <td>UniProtKB:A0A075B6H5</td>
      <td>TRBV20OR9-2</td>
      <td>involved_in</td>
      <td>GO:0002376</td>
      <td>immune system process</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0008150, BFO:0000015, GO:0002376, BFO:0000...</td>
      <td>{BFO:0000015, GO:0008150, BFO:0000003}</td>
      <td>IEA</td>
      <td>False</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000043]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>782818</th>
      <td>UniProtKB:Q9NZC2</td>
      <td>TREM2</td>
      <td>involved_in</td>
      <td>GO:0045088</td>
      <td>regulation of innate immune response</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0065007, GO:0002682, GO:0050776, BFO:00000...</td>
      <td>{GO:0065007, GO:0002682, GO:0050776, BFO:00000...</td>
      <td>IBA</td>
      <td>True</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>782819</th>
      <td>UniProtKB:Q9Y2K2</td>
      <td>SIK3</td>
      <td>enables</td>
      <td>GO:0050321</td>
      <td>tau-protein kinase activity</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0140096, GO:0003824, GO:0004674, GO:001674...</td>
      <td>{GO:0140096, GO:0003824, GO:0004674, GO:001674...</td>
      <td>IBA</td>
      <td>True</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>782820</th>
      <td>UniProtKB:P43235</td>
      <td>CTSK</td>
      <td>involved_in</td>
      <td>GO:0051603</td>
      <td>proteolysis involved in protein catabolic process</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0043170, GO:0006508, GO:0044238, GO:000905...</td>
      <td>{GO:0043170, GO:0006508, GO:0044238, GO:000905...</td>
      <td>IBA</td>
      <td>True</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>782821</th>
      <td>UniProtKB:Q07343</td>
      <td>PDE4B</td>
      <td>enables</td>
      <td>GO:0047555</td>
      <td>3',5'-cyclic-GMP phosphodiesterase activity</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0047555, GO:0003824, GO:0008081, BFO:00000...</td>
      <td>{GO:0003824, GO:0008081, BFO:0000015, GO:00425...</td>
      <td>IBA</td>
      <td>True</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
    </tr>
    <tr>
      <th>782822</th>
      <td>UniProtKB:A6NC42</td>
      <td>DPPA5</td>
      <td>involved_in</td>
      <td>GO:0010468</td>
      <td>regulation of gene expression</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0065007, GO:0060255, GO:0009889, BFO:00000...</td>
      <td>{GO:0065007, GO:0060255, GO:0009889, BFO:00000...</td>
      <td>IBA</td>
      <td>True</td>
      <td>None</td>
      <td>None</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
    </tr>
  </tbody>
</table>
<p>782823 rows × 15 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[311]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>set_iba_status(test_df)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[312]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>test_df[&#39;iba_rejected&#39;].unique()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[312]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
array([False,  True])
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[313]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>#set_iba_status(df)
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[314]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>def synthesize(grp: str) -&gt; pd.DataFrame:
    &quot;&quot;&quot;
    Synthesizes the above steps
    &quot;&quot;&quot;
    assocs = assocs_to_df(db[grp][LATEST], LATEST)
    prev_df_sets = []
    for prev in PREVIOUS:
        this_prev_assocs = assocs_to_df(db[grp][prev], prev)
        prev_df_sets.append(this_prev_assocs)
    annotate_new_pubs(assocs, prev_df_sets[0])
    #prev_assocs = pd.concat(prev_df_sets).drop_duplicates()
    prev_assocs = pd.concat(prev_df_sets)
    # prev_assocs = assocs_to_df(db[grp][PREVIOUS[0]], PREVIOUS[0])
    repair_assocs_df(assocs)
    repair_assocs_df(prev_assocs)
    # Create a gene dataframe (for latest only)
    gene_df = create_gene_df(assocs)
    annotate_redacted_pubs(assocs, prev_assocs)
    new_assocs = pd.concat([assocs, prev_assocs[prev_assocs[&#39;redacted&#39;]]])
    set_redundant_flag(new_assocs, gene_df)
    set_iba_status(new_assocs)
    return new_assocs
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[315]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>new_human = synthesize(&quot;goa_human&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Multiple IDs for 64 labels
[(&#39;AKAP7&#39;, [&#39;UniProtKB:O43687&#39;, &#39;UniProtKB:Q9P0M2&#39;]), (&#39;ARHGEF18&#39;, [&#39;UniProtKB:A0A590UK10&#39;, &#39;UniProtKB:Q6ZSZ5&#39;]), (&#39;BBC3&#39;, [&#39;UniProtKB:Q9BXH1&#39;, &#39;UniProtKB:Q96PG8&#39;]), (&#39;CALCA&#39;, [&#39;UniProtKB:P06881&#39;, &#39;UniProtKB:P01258&#39;]), (&#39;CDKN2A&#39;, [&#39;UniProtKB:Q8N726&#39;, &#39;UniProtKB:P42771&#39;])]
Multiple IDs for 100 labels
[(&#39;AKAP7&#39;, [&#39;UniProtKB:O43687&#39;, &#39;UniProtKB:Q9P0M2&#39;]), (&#39;AMY1A&#39;, [&#39;UniProtKB:P04745&#39;, &#39;UniProtKB:P0DUB6&#39;]), (&#39;ARHGEF18&#39;, [&#39;UniProtKB:A0A590UK10&#39;, &#39;UniProtKB:Q6ZSZ5&#39;]), (&#39;ASIC5&#39;, [&#39;UniProtKB:A0A0G2JLG4&#39;, &#39;UniProtKB:Q9NY37&#39;]), (&#39;ATP6AP2&#39;, [&#39;UniProtKB:O75787&#39;, &#39;UniProtKB:A0A1C7CYW4&#39;])]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[316]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>new_human
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[316]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>subject</th>
      <th>subject_label</th>
      <th>predicate</th>
      <th>object</th>
      <th>object_label</th>
      <th>object_obsoletes</th>
      <th>object_uninformative</th>
      <th>object_closure</th>
      <th>object_closure_redundant</th>
      <th>evidence</th>
      <th>...</th>
      <th>pubs</th>
      <th>release</th>
      <th>pmid_new</th>
      <th>is_new</th>
      <th>fresh</th>
      <th>pmid_removed</th>
      <th>unique</th>
      <th>redacted</th>
      <th>redundant</th>
      <th>iba_rejected</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>UniProtKB:A0A024RBG1</td>
      <td>NUDT4B</td>
      <td>enables</td>
      <td>GO:0003723</td>
      <td>RNA binding</td>
      <td>False</td>
      <td>True</td>
      <td>{GO:0097159, GO:0003723, BFO:0000015, GO:00036...</td>
      <td>{GO:0097159, BFO:0000015, GO:0003674, GO:00036...</td>
      <td>IEA</td>
      <td>...</td>
      <td>[GO_REF:0000043]</td>
      <td>2024-11-03</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>True</td>
    </tr>
    <tr>
      <th>1</th>
      <td>UniProtKB:A0A024RBG1</td>
      <td>NUDT4B</td>
      <td>enables</td>
      <td>GO:0005515</td>
      <td>protein binding</td>
      <td>False</td>
      <td>True</td>
      <td>{BFO:0000015, GO:0003674, GO:0005488, BFO:0000...</td>
      <td>{GO:0003674, BFO:0000015, GO:0005488, BFO:0000...</td>
      <td>IPI</td>
      <td>...</td>
      <td>[PMID:33961781]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>True</td>
    </tr>
    <tr>
      <th>2</th>
      <td>UniProtKB:A0A024RBG1</td>
      <td>NUDT4B</td>
      <td>enables</td>
      <td>GO:0046872</td>
      <td>metal ion binding</td>
      <td>False</td>
      <td>True</td>
      <td>{GO:0043169, GO:0046872, BFO:0000015, GO:00360...</td>
      <td>{GO:0043169, BFO:0000015, GO:0036094, GO:00431...</td>
      <td>IEA</td>
      <td>...</td>
      <td>[GO_REF:0000043]</td>
      <td>2024-11-03</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>True</td>
    </tr>
    <tr>
      <th>3</th>
      <td>UniProtKB:A0A024RBG1</td>
      <td>NUDT4B</td>
      <td>located_in</td>
      <td>GO:0005829</td>
      <td>cytosol</td>
      <td>False</td>
      <td>False</td>
      <td>{CARO:0030000, UBERON:0000061, CARO:0000003, G...</td>
      <td>{CARO:0030000, CARO:0000000, BFO:0000004, GO:0...</td>
      <td>IDA</td>
      <td>...</td>
      <td>[GO_REF:0000052]</td>
      <td>2024-11-03</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>True</td>
    </tr>
    <tr>
      <th>4</th>
      <td>UniProtKB:A0A075B6H5</td>
      <td>TRBV20OR9-2</td>
      <td>involved_in</td>
      <td>GO:0002376</td>
      <td>immune system process</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0008150, BFO:0000015, GO:0002376, BFO:0000...</td>
      <td>{BFO:0000015, GO:0008150, BFO:0000003}</td>
      <td>IEA</td>
      <td>...</td>
      <td>[GO_REF:0000043]</td>
      <td>2024-11-03</td>
      <td>False</td>
      <td>True</td>
      <td>False</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>True</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>434570</th>
      <td>UniProtKB:Q9Y6A4</td>
      <td>CFAP20</td>
      <td>None</td>
      <td>GO:0007275</td>
      <td>multicellular organism development</td>
      <td>False</td>
      <td>True</td>
      <td>{GO:0032501, BFO:0000015, GO:0048856, GO:00325...</td>
      <td>{GO:0032501, BFO:0000015, GO:0048856, GO:00325...</td>
      <td>TAS</td>
      <td>...</td>
      <td>[PMID:8688464]</td>
      <td>2020-01-01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
      <td>True</td>
    </tr>
    <tr>
      <th>435058</th>
      <td>UniProtKB:Q9Y6F1</td>
      <td>PARP3</td>
      <td>None</td>
      <td>GO:0006281</td>
      <td>DNA repair</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0043170, GO:0033554, GO:0008152, GO:000613...</td>
      <td>{GO:0043170, GO:0008152, BFO:0000015, GO:00099...</td>
      <td>TAS</td>
      <td>...</td>
      <td>[PMID:7260241]</td>
      <td>2020-01-01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
    </tr>
    <tr>
      <th>436209</th>
      <td>UniProtKB:Q9Y6Q9</td>
      <td>NCOA3</td>
      <td>None</td>
      <td>GO:0000981</td>
      <td>DNA-binding transcription factor activity, RNA...</td>
      <td>False</td>
      <td>False</td>
      <td>{BFO:0000015, GO:0006357, GO:0008150, GO:00192...</td>
      <td>{GO:0065007, GO:0060255, GO:0009889, BFO:00000...</td>
      <td>ISM</td>
      <td>...</td>
      <td>[PMID:19274049]</td>
      <td>2020-01-01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
      <td>True</td>
    </tr>
    <tr>
      <th>436550</th>
      <td>UniProtKB:Q9Y6X0</td>
      <td>SETBP1</td>
      <td>None</td>
      <td>GO:0000981</td>
      <td>DNA-binding transcription factor activity, RNA...</td>
      <td>False</td>
      <td>False</td>
      <td>{BFO:0000015, GO:0006357, GO:0008150, GO:00192...</td>
      <td>{GO:0065007, GO:0060255, GO:0009889, BFO:00000...</td>
      <td>ISM</td>
      <td>...</td>
      <td>[PMID:19274049]</td>
      <td>2020-01-01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
      <td>True</td>
    </tr>
    <tr>
      <th>436732</th>
      <td>UniProtKB:Q9Y6Y1</td>
      <td>CAMTA1</td>
      <td>None</td>
      <td>GO:0000981</td>
      <td>DNA-binding transcription factor activity, RNA...</td>
      <td>False</td>
      <td>False</td>
      <td>{BFO:0000015, GO:0006357, GO:0008150, GO:00192...</td>
      <td>{GO:0065007, GO:0060255, GO:0009889, BFO:00000...</td>
      <td>ISM</td>
      <td>...</td>
      <td>[PMID:19274049]</td>
      <td>2020-01-01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
<p>783918 rows × 23 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[317]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>new_human[new_human[&#39;iba_rejected&#39;]]
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[317]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>subject</th>
      <th>subject_label</th>
      <th>predicate</th>
      <th>object</th>
      <th>object_label</th>
      <th>object_obsoletes</th>
      <th>object_uninformative</th>
      <th>object_closure</th>
      <th>object_closure_redundant</th>
      <th>evidence</th>
      <th>...</th>
      <th>pubs</th>
      <th>release</th>
      <th>pmid_new</th>
      <th>is_new</th>
      <th>fresh</th>
      <th>pmid_removed</th>
      <th>unique</th>
      <th>redacted</th>
      <th>redundant</th>
      <th>iba_rejected</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>UniProtKB:A0A024RBG1</td>
      <td>NUDT4B</td>
      <td>enables</td>
      <td>GO:0003723</td>
      <td>RNA binding</td>
      <td>False</td>
      <td>True</td>
      <td>{GO:0097159, GO:0003723, BFO:0000015, GO:00036...</td>
      <td>{GO:0097159, BFO:0000015, GO:0003674, GO:00036...</td>
      <td>IEA</td>
      <td>...</td>
      <td>[GO_REF:0000043]</td>
      <td>2024-11-03</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>True</td>
    </tr>
    <tr>
      <th>1</th>
      <td>UniProtKB:A0A024RBG1</td>
      <td>NUDT4B</td>
      <td>enables</td>
      <td>GO:0005515</td>
      <td>protein binding</td>
      <td>False</td>
      <td>True</td>
      <td>{BFO:0000015, GO:0003674, GO:0005488, BFO:0000...</td>
      <td>{GO:0003674, BFO:0000015, GO:0005488, BFO:0000...</td>
      <td>IPI</td>
      <td>...</td>
      <td>[PMID:33961781]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>True</td>
    </tr>
    <tr>
      <th>2</th>
      <td>UniProtKB:A0A024RBG1</td>
      <td>NUDT4B</td>
      <td>enables</td>
      <td>GO:0046872</td>
      <td>metal ion binding</td>
      <td>False</td>
      <td>True</td>
      <td>{GO:0043169, GO:0046872, BFO:0000015, GO:00360...</td>
      <td>{GO:0043169, BFO:0000015, GO:0036094, GO:00431...</td>
      <td>IEA</td>
      <td>...</td>
      <td>[GO_REF:0000043]</td>
      <td>2024-11-03</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>True</td>
    </tr>
    <tr>
      <th>3</th>
      <td>UniProtKB:A0A024RBG1</td>
      <td>NUDT4B</td>
      <td>located_in</td>
      <td>GO:0005829</td>
      <td>cytosol</td>
      <td>False</td>
      <td>False</td>
      <td>{CARO:0030000, UBERON:0000061, CARO:0000003, G...</td>
      <td>{CARO:0030000, CARO:0000000, BFO:0000004, GO:0...</td>
      <td>IDA</td>
      <td>...</td>
      <td>[GO_REF:0000052]</td>
      <td>2024-11-03</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>True</td>
    </tr>
    <tr>
      <th>4</th>
      <td>UniProtKB:A0A075B6H5</td>
      <td>TRBV20OR9-2</td>
      <td>involved_in</td>
      <td>GO:0002376</td>
      <td>immune system process</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0008150, BFO:0000015, GO:0002376, BFO:0000...</td>
      <td>{BFO:0000015, GO:0008150, BFO:0000003}</td>
      <td>IEA</td>
      <td>...</td>
      <td>[GO_REF:0000043]</td>
      <td>2024-11-03</td>
      <td>False</td>
      <td>True</td>
      <td>False</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>True</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>433680</th>
      <td>UniProtKB:Q9Y5Y6</td>
      <td>ST14</td>
      <td>None</td>
      <td>GO:0005887</td>
      <td>None</td>
      <td>True</td>
      <td>False</td>
      <td>{GO:0005887}</td>
      <td>{}</td>
      <td>TAS</td>
      <td>...</td>
      <td>[PMID:10831593]</td>
      <td>2020-01-01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
      <td>True</td>
    </tr>
    <tr>
      <th>434570</th>
      <td>UniProtKB:Q9Y6A4</td>
      <td>CFAP20</td>
      <td>None</td>
      <td>GO:0007275</td>
      <td>multicellular organism development</td>
      <td>False</td>
      <td>True</td>
      <td>{GO:0032501, BFO:0000015, GO:0048856, GO:00325...</td>
      <td>{GO:0032501, BFO:0000015, GO:0048856, GO:00325...</td>
      <td>TAS</td>
      <td>...</td>
      <td>[PMID:8688464]</td>
      <td>2020-01-01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
      <td>True</td>
    </tr>
    <tr>
      <th>436209</th>
      <td>UniProtKB:Q9Y6Q9</td>
      <td>NCOA3</td>
      <td>None</td>
      <td>GO:0000981</td>
      <td>DNA-binding transcription factor activity, RNA...</td>
      <td>False</td>
      <td>False</td>
      <td>{BFO:0000015, GO:0006357, GO:0008150, GO:00192...</td>
      <td>{GO:0065007, GO:0060255, GO:0009889, BFO:00000...</td>
      <td>ISM</td>
      <td>...</td>
      <td>[PMID:19274049]</td>
      <td>2020-01-01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
      <td>True</td>
    </tr>
    <tr>
      <th>436550</th>
      <td>UniProtKB:Q9Y6X0</td>
      <td>SETBP1</td>
      <td>None</td>
      <td>GO:0000981</td>
      <td>DNA-binding transcription factor activity, RNA...</td>
      <td>False</td>
      <td>False</td>
      <td>{BFO:0000015, GO:0006357, GO:0008150, GO:00192...</td>
      <td>{GO:0065007, GO:0060255, GO:0009889, BFO:00000...</td>
      <td>ISM</td>
      <td>...</td>
      <td>[PMID:19274049]</td>
      <td>2020-01-01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
      <td>True</td>
    </tr>
    <tr>
      <th>436732</th>
      <td>UniProtKB:Q9Y6Y1</td>
      <td>CAMTA1</td>
      <td>None</td>
      <td>GO:0000981</td>
      <td>DNA-binding transcription factor activity, RNA...</td>
      <td>False</td>
      <td>False</td>
      <td>{BFO:0000015, GO:0006357, GO:0008150, GO:00192...</td>
      <td>{GO:0065007, GO:0060255, GO:0009889, BFO:00000...</td>
      <td>ISM</td>
      <td>...</td>
      <td>[PMID:19274049]</td>
      <td>2020-01-01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
<p>450057 rows × 23 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[318]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>prev_human_df = assocs_to_df(db[&quot;goa_human&quot;][PREVIOUS[0]], PREVIOUS[0])
#annotate_new_pubs(human_df, prev_human_df)
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[319]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>annotate_new_pubs(human_df, prev_human_df)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[320]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>human_df[human_df[&#39;fresh&#39;]]
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[320]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>subject</th>
      <th>subject_label</th>
      <th>predicate</th>
      <th>object</th>
      <th>object_label</th>
      <th>object_obsoletes</th>
      <th>object_uninformative</th>
      <th>object_closure</th>
      <th>object_closure_redundant</th>
      <th>evidence</th>
      <th>is_iba</th>
      <th>negated</th>
      <th>pmid</th>
      <th>pubs</th>
      <th>release</th>
      <th>pmid_new</th>
      <th>is_new</th>
      <th>fresh</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>UniProtKB:A0A024RBG1</td>
      <td>NUDT4B</td>
      <td>enables</td>
      <td>GO:0005515</td>
      <td>protein binding</td>
      <td>False</td>
      <td>True</td>
      <td>{BFO:0000015, GO:0003674, GO:0005488, BFO:0000...</td>
      <td>{GO:0003674, BFO:0000015, GO:0005488, BFO:0000...</td>
      <td>IPI</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:33961781</td>
      <td>[PMID:33961781]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
    </tr>
    <tr>
      <th>357</th>
      <td>UniProtKB:A0A096LP55</td>
      <td>UQCRHL</td>
      <td>located_in</td>
      <td>GO:0005739</td>
      <td>mitochondrion</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0005739, GO:0110165, GO:0043231, BFO:00000...</td>
      <td>{CARO:0030000, CARO:0000000, BFO:0000004, GO:0...</td>
      <td>HTP</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:34800366</td>
      <td>[PMID:34800366]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
    </tr>
    <tr>
      <th>569</th>
      <td>UniProtKB:A0A0B4J2F0</td>
      <td>PIGBOS1</td>
      <td>located_in</td>
      <td>GO:0005739</td>
      <td>mitochondrion</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0005739, GO:0110165, GO:0043231, BFO:00000...</td>
      <td>{CARO:0030000, CARO:0000000, BFO:0000004, GO:0...</td>
      <td>HTP</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:34800366</td>
      <td>[PMID:34800366]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
    </tr>
    <tr>
      <th>1904</th>
      <td>UniProtKB:A1A4Y4</td>
      <td>IRGM</td>
      <td>enables</td>
      <td>GO:1901612</td>
      <td>cardiolipin binding</td>
      <td>False</td>
      <td>True</td>
      <td>{GO:0008289, GO:0043168, GO:1901612, BFO:00000...</td>
      <td>{GO:0008289, GO:0043168, BFO:0000015, GO:00360...</td>
      <td>IDA</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:21102437</td>
      <td>[PMID:21102437]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
    </tr>
    <tr>
      <th>1945</th>
      <td>UniProtKB:A1A4Y4</td>
      <td>IRGM</td>
      <td>involved_in</td>
      <td>GO:0090141</td>
      <td>positive regulation of mitochondrial fission</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0010821, GO:0051130, GO:0065007, GO:003304...</td>
      <td>{GO:0010821, GO:0051130, GO:0065007, GO:003304...</td>
      <td>IDA</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:21102437</td>
      <td>[PMID:21102437]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>718365</th>
      <td>RNAcentral:URS00026A23F2_9606</td>
      <td>URS00026A23F2_9606</td>
      <td>enables</td>
      <td>GO:0141180</td>
      <td>dsDNA-RNA triple helix-forming chromatin adapt...</td>
      <td>False</td>
      <td>True</td>
      <td>{GO:0030674, BFO:0000015, GO:0043565, GO:00036...</td>
      <td>{GO:0003690, GO:0030674, GO:0003677, GO:007184...</td>
      <td>IDA</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:27634931</td>
      <td>[PMID:27634931]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
    </tr>
    <tr>
      <th>718366</th>
      <td>RNAcentral:URS00026A23F2_9606</td>
      <td>URS00026A23F2_9606</td>
      <td>involved_in</td>
      <td>GO:0000122</td>
      <td>negative regulation of transcription by RNA po...</td>
      <td>False</td>
      <td>False</td>
      <td>{BFO:0000015, GO:0006357, GO:0008150, GO:00098...</td>
      <td>{GO:0045934, GO:0009892, GO:0065007, GO:000988...</td>
      <td>IMP</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:27634931</td>
      <td>[PMID:27634931]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
    </tr>
    <tr>
      <th>718367</th>
      <td>RNAcentral:URS00026A23F2_9606</td>
      <td>URS00026A23F2_9606</td>
      <td>involved_in</td>
      <td>GO:0000512</td>
      <td>lncRNA-mediated post-transcriptional gene sile...</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0000512, BFO:0000015, GO:0016441, GO:00081...</td>
      <td>{GO:0065007, GO:0009892, GO:0031047, GO:000988...</td>
      <td>IDA</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:30720199</td>
      <td>[PMID:30720199]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
    </tr>
    <tr>
      <th>718368</th>
      <td>RNAcentral:URS00026A23F2_9606</td>
      <td>URS00026A23F2_9606</td>
      <td>involved_in</td>
      <td>GO:0000512</td>
      <td>lncRNA-mediated post-transcriptional gene sile...</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0000512, BFO:0000015, GO:0016441, GO:00081...</td>
      <td>{GO:0065007, GO:0009892, GO:0031047, GO:000988...</td>
      <td>IMP</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:33102210</td>
      <td>[PMID:33102210]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
    </tr>
    <tr>
      <th>718370</th>
      <td>RNAcentral:URS00026A23F2_9606</td>
      <td>URS00026A23F2_9606</td>
      <td>involved_in</td>
      <td>GO:0090399</td>
      <td>replicative senescence</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0008152, BFO:0000015, GO:0090399, GO:00099...</td>
      <td>{GO:0008152, BFO:0000015, GO:0009987, GO:00081...</td>
      <td>IMP</td>
      <td>False</td>
      <td>None</td>
      <td>PMID:27634931</td>
      <td>[PMID:27634931]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
<p>3380 rows × 18 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[321]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>#pair_assocs, pair_anns = annotate_assocs(&quot;tair&quot;)
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[322]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>new_human.to_csv(&quot;output/go-human-assocs-annotated.csv&quot;, index=False)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[332]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>test_cases_df = new_human[(new_human[&#39;fresh&#39;] == True) &amp; (new_human[&#39;object_uninformative&#39;] == False) &amp; (new_human[&#39;negated&#39;] != True) &amp; (new_human[&#39;redundant&#39;] != True) &amp; (new_human[&#39;iba_rejected&#39;] != True)]
test_cases_df
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[332]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>subject</th>
      <th>subject_label</th>
      <th>predicate</th>
      <th>object</th>
      <th>object_label</th>
      <th>object_obsoletes</th>
      <th>object_uninformative</th>
      <th>object_closure</th>
      <th>object_closure_redundant</th>
      <th>evidence</th>
      <th>...</th>
      <th>pubs</th>
      <th>release</th>
      <th>pmid_new</th>
      <th>is_new</th>
      <th>fresh</th>
      <th>pmid_removed</th>
      <th>unique</th>
      <th>redacted</th>
      <th>redundant</th>
      <th>iba_rejected</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>4876</th>
      <td>UniProtKB:A6NNL5</td>
      <td>C15orf61</td>
      <td>located_in</td>
      <td>GO:0005739</td>
      <td>mitochondrion</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0005739, GO:0110165, GO:0043231, BFO:00000...</td>
      <td>{CARO:0030000, CARO:0000000, BFO:0000004, GO:0...</td>
      <td>HTP</td>
      <td>...</td>
      <td>[PMID:34800366]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>5853</th>
      <td>UniProtKB:A8MSI8</td>
      <td>LYRM9</td>
      <td>located_in</td>
      <td>GO:0005739</td>
      <td>mitochondrion</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0005739, GO:0110165, GO:0043231, BFO:00000...</td>
      <td>{CARO:0030000, CARO:0000000, BFO:0000004, GO:0...</td>
      <td>HTP</td>
      <td>...</td>
      <td>[PMID:34800366]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>6327</th>
      <td>UniProtKB:A8MXV4</td>
      <td>NUDT19</td>
      <td>located_in</td>
      <td>GO:0005739</td>
      <td>mitochondrion</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0005739, GO:0110165, GO:0043231, BFO:00000...</td>
      <td>{CARO:0030000, CARO:0000000, BFO:0000004, GO:0...</td>
      <td>HTP</td>
      <td>...</td>
      <td>[PMID:34800366]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>15942</th>
      <td>UniProtKB:O14521</td>
      <td>SDHD</td>
      <td>part_of</td>
      <td>GO:0045273</td>
      <td>respiratory chain complex II (succinate dehydr...</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0098796, GO:0110165, GO:0045273, BFO:00000...</td>
      <td>{CARO:0030000, GO:0098803, GO:0098796, CARO:00...</td>
      <td>IDA</td>
      <td>...</td>
      <td>[PMID:37098072]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>34482</th>
      <td>UniProtKB:O43325</td>
      <td>LYRM1</td>
      <td>located_in</td>
      <td>GO:0005739</td>
      <td>mitochondrion</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0005739, GO:0110165, GO:0043231, BFO:00000...</td>
      <td>{CARO:0030000, CARO:0000000, BFO:0000004, GO:0...</td>
      <td>HTP</td>
      <td>...</td>
      <td>[PMID:34800366]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>718362</th>
      <td>RNAcentral:URS00026A23F2_9606</td>
      <td>URS00026A23F2_9606</td>
      <td>acts_upstream_of</td>
      <td>GO:0008284</td>
      <td>positive regulation of cell population prolife...</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0065007, GO:0042127, GO:0048518, BFO:00000...</td>
      <td>{GO:0065007, GO:0042127, GO:0048518, BFO:00000...</td>
      <td>IMP</td>
      <td>...</td>
      <td>[PMID:33102210]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>718366</th>
      <td>RNAcentral:URS00026A23F2_9606</td>
      <td>URS00026A23F2_9606</td>
      <td>involved_in</td>
      <td>GO:0000122</td>
      <td>negative regulation of transcription by RNA po...</td>
      <td>False</td>
      <td>False</td>
      <td>{BFO:0000015, GO:0006357, GO:0008150, GO:00098...</td>
      <td>{GO:0045934, GO:0009892, GO:0065007, GO:000988...</td>
      <td>IMP</td>
      <td>...</td>
      <td>[PMID:27634931]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>718367</th>
      <td>RNAcentral:URS00026A23F2_9606</td>
      <td>URS00026A23F2_9606</td>
      <td>involved_in</td>
      <td>GO:0000512</td>
      <td>lncRNA-mediated post-transcriptional gene sile...</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0000512, BFO:0000015, GO:0016441, GO:00081...</td>
      <td>{GO:0065007, GO:0009892, GO:0031047, GO:000988...</td>
      <td>IDA</td>
      <td>...</td>
      <td>[PMID:30720199]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>718368</th>
      <td>RNAcentral:URS00026A23F2_9606</td>
      <td>URS00026A23F2_9606</td>
      <td>involved_in</td>
      <td>GO:0000512</td>
      <td>lncRNA-mediated post-transcriptional gene sile...</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0000512, BFO:0000015, GO:0016441, GO:00081...</td>
      <td>{GO:0065007, GO:0009892, GO:0031047, GO:000988...</td>
      <td>IMP</td>
      <td>...</td>
      <td>[PMID:33102210]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>718370</th>
      <td>RNAcentral:URS00026A23F2_9606</td>
      <td>URS00026A23F2_9606</td>
      <td>involved_in</td>
      <td>GO:0090399</td>
      <td>replicative senescence</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0008152, BFO:0000015, GO:0090399, GO:00099...</td>
      <td>{GO:0008152, BFO:0000015, GO:0009987, GO:00081...</td>
      <td>IMP</td>
      <td>...</td>
      <td>[PMID:27634931]</td>
      <td>2024-11-03</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
  </tbody>
</table>
<p>524 rows × 23 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[350]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>def row_to_test_case(row, answer=&quot;YES&quot;):
    gene = row[&#39;subject_label&#39;]
    term = row[&#39;object_label&#39;]
    predicate = row[&#39;predicate&#39;]
    if not gene or not term or not predicate:
        return

    return {
        &quot;input&quot;: f&quot;{gene} {predicate} {term}&quot;,
        &quot;original_input&quot;: {
            &quot;subject&quot;: str(row[&#39;subject&#39;]),
            &quot;predicate&quot;: str(row[&#39;predicate&#39;]),
            &quot;object&quot;: str(row[&#39;object&#39;]),
        },
        &quot;ideal&quot;: answer,
    }


def df_to_test_cases(df: pd.DataFrame, limit=1000):
    cases = [row_to_test_case(row) for _, row in df.iterrows()]
    cases = [x for x in cases if x is not None]
    if limit:
        cases = cases[:limit]
    return cases

df_to_test_cases(test_cases_df, limit=5)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[350]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[{&#39;input&#39;: &#39;C15orf61 located_in mitochondrion&#39;,
  &#39;original_input&#39;: {&#39;subject&#39;: &#39;UniProtKB:A6NNL5&#39;,
   &#39;predicate&#39;: &#39;located_in&#39;,
   &#39;object&#39;: &#39;GO:0005739&#39;},
  &#39;ideal&#39;: &#39;YES&#39;},
 {&#39;input&#39;: &#39;LYRM9 located_in mitochondrion&#39;,
  &#39;original_input&#39;: {&#39;subject&#39;: &#39;UniProtKB:A8MSI8&#39;,
   &#39;predicate&#39;: &#39;located_in&#39;,
   &#39;object&#39;: &#39;GO:0005739&#39;},
  &#39;ideal&#39;: &#39;YES&#39;},
 {&#39;input&#39;: &#39;NUDT19 located_in mitochondrion&#39;,
  &#39;original_input&#39;: {&#39;subject&#39;: &#39;UniProtKB:A8MXV4&#39;,
   &#39;predicate&#39;: &#39;located_in&#39;,
   &#39;object&#39;: &#39;GO:0005739&#39;},
  &#39;ideal&#39;: &#39;YES&#39;},
 {&#39;input&#39;: &#39;SDHD part_of respiratory chain complex II (succinate dehydrogenase)&#39;,
  &#39;original_input&#39;: {&#39;subject&#39;: &#39;UniProtKB:O14521&#39;,
   &#39;predicate&#39;: &#39;part_of&#39;,
   &#39;object&#39;: &#39;GO:0045273&#39;},
  &#39;ideal&#39;: &#39;YES&#39;},
 {&#39;input&#39;: &#39;LYRM1 located_in mitochondrion&#39;,
  &#39;original_input&#39;: {&#39;subject&#39;: &#39;UniProtKB:O43325&#39;,
   &#39;predicate&#39;: &#39;located_in&#39;,
   &#39;object&#39;: &#39;GO:0005739&#39;},
  &#39;ideal&#39;: &#39;YES&#39;}]
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[341]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>#cases = [row_to_test_case(row) for _, row in test_cases_df.iterrows()]
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[343]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>import yaml

with open(&quot;output/test-cases.yaml&quot;, &quot;w&quot;) as f:
    yaml.dump({&quot;cases&quot;: cases}, f, sort_keys=False)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[345]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>redacted_df = new_human[(new_human[&#39;redacted&#39;] == True) &amp; (new_human[&#39;object_uninformative&#39;] == False) &amp; (new_human[&#39;negated&#39;] != True) &amp; (new_human[&#39;redundant&#39;] != True) &amp; (new_human[&#39;iba_rejected&#39;] != True)]
redacted_df
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[345]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>subject</th>
      <th>subject_label</th>
      <th>predicate</th>
      <th>object</th>
      <th>object_label</th>
      <th>object_obsoletes</th>
      <th>object_uninformative</th>
      <th>object_closure</th>
      <th>object_closure_redundant</th>
      <th>evidence</th>
      <th>...</th>
      <th>pubs</th>
      <th>release</th>
      <th>pmid_new</th>
      <th>is_new</th>
      <th>fresh</th>
      <th>pmid_removed</th>
      <th>unique</th>
      <th>redacted</th>
      <th>redundant</th>
      <th>iba_rejected</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>401144</th>
      <td>UniProtKB:Q8N6R0</td>
      <td>METTL13</td>
      <td>involved_in</td>
      <td>GO:0000122</td>
      <td>negative regulation of transcription by RNA po...</td>
      <td>False</td>
      <td>False</td>
      <td>{BFO:0000015, GO:0006357, GO:0008150, GO:00098...</td>
      <td>{GO:0045934, GO:0009892, GO:0065007, GO:000988...</td>
      <td>IMP</td>
      <td>...</td>
      <td>[PMID:26763933]</td>
      <td>2024-06-10</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>401146</th>
      <td>UniProtKB:Q8N6R0</td>
      <td>METTL13</td>
      <td>involved_in</td>
      <td>GO:1902807</td>
      <td>negative regulation of cell cycle G1/S phase t...</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0065007, GO:1902806, GO:0010948, GO:190198...</td>
      <td>{GO:0065007, GO:0010948, GO:1901988, GO:005172...</td>
      <td>IMP</td>
      <td>...</td>
      <td>[PMID:26763933]</td>
      <td>2024-06-10</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>453842</th>
      <td>UniProtKB:Q96K19</td>
      <td>RNF170</td>
      <td>involved_in</td>
      <td>GO:0034140</td>
      <td>negative regulation of toll-like receptor 3 si...</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0048585, GO:0062207, GO:1902532, GO:000996...</td>
      <td>{GO:0048585, GO:0065007, GO:0002682, GO:006220...</td>
      <td>IDA</td>
      <td>...</td>
      <td>[PMID:31076723]</td>
      <td>2024-06-10</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>587451</th>
      <td>RNAcentral:URS0000083D87_9606</td>
      <td>URS0000083D87_9606</td>
      <td>involved_in</td>
      <td>GO:0035195</td>
      <td>miRNA-mediated post-transcriptional gene silen...</td>
      <td>False</td>
      <td>False</td>
      <td>{BFO:0000015, GO:0016441, GO:0008150, GO:00106...</td>
      <td>{GO:0065007, GO:0009892, GO:0031047, GO:000988...</td>
      <td>IDA</td>
      <td>...</td>
      <td>[PMID:28640956]</td>
      <td>2024-06-10</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>587452</th>
      <td>RNAcentral:URS0000083D87_9606</td>
      <td>URS0000083D87_9606</td>
      <td>involved_in</td>
      <td>GO:0090051</td>
      <td>negative regulation of cell migration involved...</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0030336, BFO:0000015, GO:0008150, GO:00105...</td>
      <td>{GO:0065007, GO:0030336, BFO:0000015, GO:00485...</td>
      <td>IGI</td>
      <td>...</td>
      <td>[PMID:28640956]</td>
      <td>2024-06-10</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>424250</th>
      <td>UniProtKB:Q9Y226</td>
      <td>SLC22A13</td>
      <td>None</td>
      <td>GO:0015695</td>
      <td>organic cation transport</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0051234, GO:0006810, BFO:0000015, GO:00156...</td>
      <td>{GO:0051234, GO:0006810, BFO:0000015, GO:00511...</td>
      <td>NAS</td>
      <td>...</td>
      <td>[PMID:10072596]</td>
      <td>2020-01-01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>425053</th>
      <td>UniProtKB:Q9Y267</td>
      <td>SLC22A14</td>
      <td>None</td>
      <td>GO:0005887</td>
      <td>None</td>
      <td>True</td>
      <td>False</td>
      <td>{GO:0005887}</td>
      <td>{}</td>
      <td>NAS</td>
      <td>...</td>
      <td>[PMID:10072596]</td>
      <td>2020-01-01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>425054</th>
      <td>UniProtKB:Q9Y267</td>
      <td>SLC22A14</td>
      <td>None</td>
      <td>GO:0015101</td>
      <td>organic cation transmembrane transporter activity</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0051234, GO:0055085, GO:0006810, BFO:00000...</td>
      <td>{GO:0051234, GO:0055085, GO:0006810, BFO:00000...</td>
      <td>NAS</td>
      <td>...</td>
      <td>[PMID:10072596]</td>
      <td>2020-01-01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>425055</th>
      <td>UniProtKB:Q9Y267</td>
      <td>SLC22A14</td>
      <td>None</td>
      <td>GO:0015695</td>
      <td>organic cation transport</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0051234, GO:0006810, BFO:0000015, GO:00156...</td>
      <td>{GO:0051234, GO:0006810, BFO:0000015, GO:00511...</td>
      <td>NAS</td>
      <td>...</td>
      <td>[PMID:10072596]</td>
      <td>2020-01-01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>432681</th>
      <td>UniProtKB:Q9Y5M6</td>
      <td>OCLM</td>
      <td>None</td>
      <td>GO:0007601</td>
      <td>visual perception</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0032501, GO:0050953, GO:0003008, BFO:00000...</td>
      <td>{GO:0032501, GO:0050953, GO:0003008, BFO:00000...</td>
      <td>TAS</td>
      <td>...</td>
      <td>[PMID:10362512]</td>
      <td>2020-01-01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>False</td>
      <td>False</td>
    </tr>
  </tbody>
</table>
<p>83 rows × 23 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[346]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>cases = [row_to_test_case(row) for _, row in redacted_df.iterrows()]

with open(&quot;output/test-cases-redacted.yaml&quot;, &quot;w&quot;) as f:
    yaml.dump({&quot;cases&quot;: cases}, f, sort_keys=False)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[348]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>iba_df = new_human[(new_human[&#39;is_iba&#39;] == True) &amp; (new_human[&#39;object_uninformative&#39;] == False) &amp; (new_human[&#39;negated&#39;] != True) &amp; (new_human[&#39;redundant&#39;] != True)]
iba_df
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[348]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>subject</th>
      <th>subject_label</th>
      <th>predicate</th>
      <th>object</th>
      <th>object_label</th>
      <th>object_obsoletes</th>
      <th>object_uninformative</th>
      <th>object_closure</th>
      <th>object_closure_redundant</th>
      <th>evidence</th>
      <th>...</th>
      <th>pubs</th>
      <th>release</th>
      <th>pmid_new</th>
      <th>is_new</th>
      <th>fresh</th>
      <th>pmid_removed</th>
      <th>unique</th>
      <th>redacted</th>
      <th>redundant</th>
      <th>iba_rejected</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>718564</th>
      <td>UniProtKB:Q06418</td>
      <td>TYRO3</td>
      <td>enables</td>
      <td>GO:0004714</td>
      <td>transmembrane receptor protein tyrosine kinase...</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0019199, GO:0140096, GO:0003824, GO:001674...</td>
      <td>{GO:0019199, GO:0003824, BFO:0000015, GO:00167...</td>
      <td>IBA</td>
      <td>...</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>718565</th>
      <td>UniProtKB:P78559</td>
      <td>MAP1A</td>
      <td>is_active_in</td>
      <td>GO:0030425</td>
      <td>dendrite</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0030425, GO:0120025, GO:0043005, GO:011016...</td>
      <td>{CARO:0030000, CL:0002319, CL:0000211, UBERON:...</td>
      <td>IBA</td>
      <td>...</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>718566</th>
      <td>UniProtKB:Q7L1W4</td>
      <td>LRRC8D</td>
      <td>is_active_in</td>
      <td>GO:0005737</td>
      <td>cytoplasm</td>
      <td>False</td>
      <td>False</td>
      <td>{CARO:0030000, UBERON:0000061, CARO:0000003, G...</td>
      <td>{CARO:0030000, UBERON:0000061, CARO:0000003, C...</td>
      <td>IBA</td>
      <td>...</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>718567</th>
      <td>UniProtKB:A3QJZ7</td>
      <td>PRAMEF27</td>
      <td>part_of</td>
      <td>GO:0031462</td>
      <td>Cul2-RING ubiquitin ligase complex</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0031462, GO:0031461, GO:1990234, BFO:00000...</td>
      <td>{GO:0031461, GO:1990234, BFO:0000004, GO:01405...</td>
      <td>IBA</td>
      <td>...</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
      <td>False</td>
      <td>True</td>
      <td>False</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>718568</th>
      <td>UniProtKB:Q70IA6</td>
      <td>MOB2</td>
      <td>is_active_in</td>
      <td>GO:0005634</td>
      <td>nucleus</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0110165, GO:0043231, BFO:0000002, GO:00432...</td>
      <td>{CARO:0030000, CARO:0000000, BFO:0000004, GO:0...</td>
      <td>IBA</td>
      <td>...</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>782818</th>
      <td>UniProtKB:Q9NZC2</td>
      <td>TREM2</td>
      <td>involved_in</td>
      <td>GO:0045088</td>
      <td>regulation of innate immune response</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0065007, GO:0002682, GO:0050776, BFO:00000...</td>
      <td>{GO:0065007, GO:0002682, GO:0050776, BFO:00000...</td>
      <td>IBA</td>
      <td>...</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>782819</th>
      <td>UniProtKB:Q9Y2K2</td>
      <td>SIK3</td>
      <td>enables</td>
      <td>GO:0050321</td>
      <td>tau-protein kinase activity</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0140096, GO:0003824, GO:0004674, GO:001674...</td>
      <td>{GO:0140096, GO:0003824, GO:0004674, GO:001674...</td>
      <td>IBA</td>
      <td>...</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>782820</th>
      <td>UniProtKB:P43235</td>
      <td>CTSK</td>
      <td>involved_in</td>
      <td>GO:0051603</td>
      <td>proteolysis involved in protein catabolic process</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0043170, GO:0006508, GO:0044238, GO:000905...</td>
      <td>{GO:0043170, GO:0006508, GO:0044238, GO:000905...</td>
      <td>IBA</td>
      <td>...</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>782821</th>
      <td>UniProtKB:Q07343</td>
      <td>PDE4B</td>
      <td>enables</td>
      <td>GO:0047555</td>
      <td>3',5'-cyclic-GMP phosphodiesterase activity</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0047555, GO:0003824, GO:0008081, BFO:00000...</td>
      <td>{GO:0003824, GO:0008081, BFO:0000015, GO:00425...</td>
      <td>IBA</td>
      <td>...</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>782822</th>
      <td>UniProtKB:A6NC42</td>
      <td>DPPA5</td>
      <td>involved_in</td>
      <td>GO:0010468</td>
      <td>regulation of gene expression</td>
      <td>False</td>
      <td>False</td>
      <td>{GO:0065007, GO:0060255, GO:0009889, BFO:00000...</td>
      <td>{GO:0065007, GO:0060255, GO:0009889, BFO:00000...</td>
      <td>IBA</td>
      <td>...</td>
      <td>[GO_REF:0000033]</td>
      <td>2024-11-03</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>False</td>
      <td>False</td>
    </tr>
  </tbody>
</table>
<p>56487 rows × 23 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[351]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>with open(&quot;output/test-cases-iba.yaml&quot;, &quot;w&quot;) as f:
    yaml.dump({&quot;cases&quot;: df_to_test_cases(iba_df, limit=1000)}, f, sort_keys=False)
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022-2026, OAK developers.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>